{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from anndata import AnnData\n",
    "import anndata\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from tifffile import imread\n",
    "from scipy.ndimage import gaussian_filter\n",
    "from scipy.sparse import csr_matrix, isspmatrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_exp0 = anndata.read_h5ad('C:/Users/mdichgan/Documents/Helmholtz/send_to_Jakob/spatial/counts_CPc_exp0_BA28.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Yao = anndata.read_h5ad(\n",
    "    'C:/Users/mdichgan/Documents/Helmholtz/send_to_Jakob/sc/Yao_150kcells_subsample_with_annotations_sparse_subset.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Yao.obs[\"celltype\"] = adata_Yao.obs[\"label\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_negative_marker_dict(adata_sp: AnnData, adata_sc: AnnData, key: str='celltype'):\n",
    "    \"\"\"add dictionary of negative marker genes for different celltypes\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    key : str\n",
    "        Celltype key in adata_sp.obs and adata_sc.obs\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    neg_marker_dict : dict[string: list]\n",
    "        dictionary with celltypes as keys and corresponding negative marker genes as values\n",
    "    \"\"\"\n",
    "\n",
    "    # Set threshold parameters\n",
    "    min_number_cells=10 # minimum number of cells belonging to a cluster to consider it in the analysis\n",
    "    max_ratio_cells=0.005 # maximum ratio of cells expressing a marker to call it a negative marker gene-ct pair\n",
    "\n",
    "    #check that genes in spatial data is subset of genes in single cell data\n",
    "    adata_sp = adata_sp[:,adata_sp.var_names.isin(adata_sc.var_names)]\n",
    "\n",
    "    # Subset adata_sc to genes of spatial data\n",
    "    adata_sc = adata_sc[:,adata_sp.var_names]           \n",
    "    \n",
    "    adata_sc.layers[\"raw\"] = adata_sc.X\n",
    "    # TMP fix for sparse matrices, ideally we don't convert, and instead have calculations for sparse/non-sparse\n",
    "    \n",
    "    if isspmatrix(adata_sc.layers[\"raw\"]):                                   #existiert dieser layer immer ->X?\n",
    "        adata_sc.layers[\"raw\"] = adata_sc.layers[\"raw\"].toarray()\n",
    "\n",
    "    # Get cell types that we find in both modalities\n",
    "    shared_celltypes = adata_sc.obs.loc[adata_sc.obs[key].isin(adata_sp.obs[key]),key].unique()\n",
    "    \n",
    "    # Filter cell types by minimum number of cells\n",
    "    celltype_count_sc = adata_sc.obs[key].value_counts().loc[shared_celltypes]\n",
    "    celltype_count_sp = adata_sc.obs[key].value_counts().loc[shared_celltypes]   \n",
    "    ct_filter = (celltype_count_sc >= min_number_cells) & (celltype_count_sp >= min_number_cells)\n",
    "    celltypes = celltype_count_sc.loc[ct_filter].index.tolist()             \n",
    "    \n",
    "    \n",
    "    # Filter cells to eligible cell types\n",
    "    adata_sc = adata_sc[adata_sc.obs[key].isin(celltypes)]\n",
    "    genes = adata_sc.var_names\n",
    "    \n",
    "    # Get ratio of positive cells per cell type\n",
    "    pos_exp_sc = pd.DataFrame(adata_sc.layers[\"raw\"] > 0,columns=adata_sp.var_names)     \n",
    "    pos_exp_sc['celltype'] = list(adata_sc.obs[key])\n",
    "    ratio_celltype_sc = pos_exp_sc.groupby('celltype').mean()\n",
    "    \n",
    "    # Get gene-cell type pairs with negative marker expression\n",
    "    neg_marker_mask = np.array(ratio_celltype_sc < max_ratio_cells)\n",
    "    \n",
    "\n",
    "    neg_marker_mask = pd.DataFrame(neg_marker_mask, columns = genes, index = ratio_celltype_sc.index)   \n",
    "    neg_marker_dict = {celltype: [] for celltype in celltypes}\n",
    "\n",
    "    for celltype, row in neg_marker_mask.iterrows():\n",
    "        for gene in genes:\n",
    "            if row[gene]:\n",
    "                neg_marker_dict[celltype].append(gene)\n",
    "\n",
    "    return neg_marker_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test get_negative_marker_dict\n",
    "dict_exp0 = get_negative_marker_dict(adata_sp = adata_exp0, adata_sc = adata_Yao, key ='celltype')\n",
    "len(dict_exp0.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_neg_mark_read_in_corr_ct(adata_sp: AnnData, adata_sc: AnnData):\n",
    "\n",
    "    \"\"\"add column in_correct_celltype and \n",
    "\n",
    "       in_correct_celltype is TRUE, if gene is in_correct_celltype, else FALSE. \n",
    "\n",
    "       Parameters\n",
    "       ----------\n",
    "       adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "       adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    \n",
    "       Returns\n",
    "       ----------\n",
    "       pandas DataFrame\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    df = adata_sp.uns[\"spots\"]  \n",
    "   \n",
    "    neg_marker_dict = get_negative_marker_dict(adata_sp, adata_sc)\n",
    "\n",
    "    celltypes = list(neg_marker_dict.keys())   \n",
    "\n",
    "    df[\"in_correct_celltype\"] = \"Unknown\"\n",
    "    df.loc[df[\"celltype\"].isin(celltypes), \"in_correct_celltype\"] = \"True\"\n",
    "\n",
    "\n",
    "    # df = df.loc[df[\"celltype\"].isin(celltypes)]         #alternative: diese zellen einfach filtern? dann spart man sich die zweite spalte\n",
    "\n",
    "    for ct in celltypes:\n",
    "      df.loc[(df[\"Gene\"].isin(neg_marker_dict[ct])) & (df[\"celltype\"] == ct), 'in_correct_celltype'] = \"False\"      \n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test get_neg_mark_read_in_corr_ct\n",
    "df_exp0 = get_neg_mark_read_in_corr_ct(adata_sp=adata_exp0, adata_sc=adata_Yao)\n",
    "print(df_exp0['x'].min(), df_exp0['x'].max(), df_exp0['y'].min(), df_exp0['y'].max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exp0[\"in_correct_celltype\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_min, x_max, y_min, y_max = df_exp0['x'].min(), df_exp0['x'].max(), df_exp0['y'].min(), df_exp0['y'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spots(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, show_ticks: bool = False):\n",
    "    \"\"\"plot gene spots\n",
    "     \n",
    "     spot is red if entry in 'in_correct_celltype' is False, blue if True, grey if unknown\n",
    "\n",
    "     Parameters\n",
    "     ----------\n",
    "     spots : pd.DataFrame\n",
    "          pd.DataFrame with columns: gene, x, y, celltype, in_correct_celltype\n",
    "     x_min : int, x_max : int, y_min : int, y_max : int \n",
    "          crop coordinates\n",
    "     image : NDArray\n",
    "          read from image of dapi stained cell-nuclei\n",
    "     show_ticks : bool \n",
    "          default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    df = spots\n",
    "    \n",
    "    s_factor = 0.6            # 15.000/((x_max-x_min)**2)                    #richtig wÃ¤hlen(x_min,...)\n",
    "    \n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    ax.axis([x_min, x_max, y_max, y_min])        \n",
    "    ax.imshow(image,cmap = \"binary_r\") \n",
    "\n",
    "    #filter spots\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "    \n",
    "    if show_ticks:\n",
    "        ax.tick_params(which='both', direction='in')\n",
    "    else:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)\n",
    "\n",
    "    ax.scatter(df.loc[df['in_correct_celltype']==\"Unknown\",\"x\"],df.loc[df['in_correct_celltype']==\"Unknown\",\"y\"], s = 0.3*s_factor, color = \"grey\")\n",
    "    ax.scatter(df.loc[df['in_correct_celltype']==\"True\",\"x\"],df.loc[df['in_correct_celltype']==\"True\",\"y\"], s = 0.5*s_factor, color = \"blue\")\n",
    "    ax.scatter(df.loc[df['in_correct_celltype']==\"False\",\"x\"],df.loc[df['in_correct_celltype']==\"False\",\"y\"], s = 1*s_factor, color = \"red\")\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "image = imread(r\"C:\\Users\\mdichgan\\Documents\\Helmholtz\\send_to_Jakob\\tif_images\\background.tiff\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_spots(df_exp0, 11500,12000,11500,12000,image,show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_neg_spot_density(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"get density array of negative spots\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, in_correct_celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of negative marker reads per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "     \n",
    "    if (x_min is None or x_max is None or y_min is None or y_max is None) and image is None:\n",
    "        raise ValueError(\"please provide an image or crop\")         \n",
    "        \n",
    "    if image is None:\n",
    "        range = [[x_min,x_max],[y_min,y_max]]\n",
    "    \n",
    "    else:\n",
    "        range = [[0,image.shape[0]],[0,image.shape[1]]]\n",
    "    \n",
    "    np.seterr(invalid='ignore')\n",
    "    #filter spots\n",
    "    df = spots\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "    true_spots = df.loc[df['in_correct_celltype'] == \"True\"]\n",
    "    x_t = true_spots['x']\n",
    "    y_t = true_spots['y']\n",
    "    H_t, x_edges, y_edges = np.histogram2d(x_t, y_t, bins = bins, range = range)\n",
    "    # Histogram does not follow Cartesian convention,\n",
    "    # therefore transpose H for visualization purposes.\n",
    "    H_t = H_t.T\n",
    "\n",
    "    false_spots = df.loc[df['in_correct_celltype'] == \"False\"]\n",
    "    x_f = false_spots['x']\n",
    "    y_f = false_spots['y']\n",
    "    H_f, x_edges, y_edges = np.histogram2d(x_f, y_f, bins = bins, range = range)\n",
    "    H_f = H_f.T\n",
    "\n",
    "\n",
    "    H = H_f/(H_t+H_f)    #ignore \"Unknown\" spots in density calculation, since it could be both True or False\n",
    "    H[np.isnan(H)] = 0      #no negative marker wrongly assigned where no spots\n",
    "\n",
    " \n",
    "    return H,range\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_neg_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_neg_spot_density(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                       image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \n",
    "    \"\"\"plot wrong spot density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, in_correct_celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    H, range = get_neg_spot_density(spots,x_min,x_max,y_min,y_max,image,bins)\n",
    "    H = gaussian_filter(H,sigma=smooth)\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(title='false spots density')\n",
    "    plot = plt.imshow(H, interpolation='nearest', origin='upper',\n",
    "            extent=[x_min, x_max, y_max, y_min], vmin=0,vmax=1)\n",
    "    fig.colorbar(plot)\n",
    "\n",
    "    if show_ticks:\n",
    "        ax.tick_params(which='both', direction='in')\n",
    "    else:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_neg_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 50, smooth=1)\n",
    "plot_neg_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 50, show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spot_density(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"get density of spots that are in celltypes of negative_marker_dict\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, in_correct_celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of spots reads per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "     \n",
    "    if (x_min is None or x_max is None or y_min is None or y_max is None) and image is None:\n",
    "        raise ValueError(\"please provide an image or crop\")         \n",
    "        \n",
    "    if image is None:\n",
    "        range = [[x_min,x_max],[y_min,y_max]]\n",
    "    \n",
    "    else:\n",
    "        range = [[0,image.shape[0]],[0,image.shape[1]]]\n",
    "    \n",
    "\n",
    "    #filter spots\n",
    "    df = spots\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "    true_spots = df.loc[df['in_correct_celltype'] != \"Unknown\"]\n",
    "    x = spots['x']\n",
    "    y = spots['y']\n",
    "    H, x_edges, y_edges = np.histogram2d(x, y, bins = bins, range = range)\n",
    "    H = H.T\n",
    "\n",
    "    return H, range\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spot_density(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"plot spot density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, in_correct_celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    H, range = get_spot_density(spots,x_min,x_max,y_min,y_max,image,bins)\n",
    "    H = gaussian_filter(H,sigma=smooth)\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(title='spots density') \n",
    "    plot = plt.imshow(H, interpolation='nearest',\n",
    "            extent=[x_min, x_max, y_max, y_min], #norm= mpl.colors.Normalize.__call__(self,value=H)\n",
    "            )\n",
    "    fig.colorbar(plot)\n",
    "\n",
    "    if show_ticks:\n",
    "        ax.tick_params(which='both', direction='in')\n",
    "    else:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 50, show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cells_df(spots: pd.DataFrame):\n",
    "    \"\"\"get cell location DataFrame\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, in_correct_celltype\n",
    "    Returns\n",
    "    -------\n",
    "        DataFrame with columns: cell, celltype, x, y\n",
    "    \"\"\"\n",
    "    df_cells = spots.loc[spots[\"in_correct_celltype\"]==\"True\"]\n",
    "    df_cells = df_cells.groupby([\"cell\",\"celltype\"])[[\"x\",\"y\"]].mean()\n",
    "    df_cells = df_cells.reset_index().rename(columns={'cell':'cell', 'celltype':'celltype'})\n",
    "\n",
    "\n",
    "    return df_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cells = get_cells_df(df_exp0)\n",
    "df_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_cells  \n",
    "x_min, x_max, y_max, y_min = 11000, 12000, 12000, 11000\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.axis([x_min, x_max, y_max, y_min])        \n",
    "ax.imshow(image,cmap = \"binary_r\") \n",
    "\n",
    "#filter spots\n",
    "df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "ax.scatter(df['x'], df['y'], s = 0.6, color = \"red\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cell_density(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"get cell density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of cells per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "     \n",
    "    if (x_min is None or x_max is None or y_min is None or y_max is None) and image is None:\n",
    "        raise ValueError(\"please provide an image or crop\")         \n",
    "        \n",
    "    if image is None:\n",
    "        range = [[x_min,x_max],[y_min,y_max]]\n",
    "    \n",
    "    else:\n",
    "        range = [[0,image.shape[0]],[0,image.shape[1]]]\n",
    "    \n",
    "\n",
    "    #filter cells\n",
    "    df = df_cells\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]    #unnÃ¶tig?\n",
    "\n",
    "    x = df['x']\n",
    "    y = df['y']\n",
    "    H, x_edges, y_edges = np.histogram2d(x, y, bins = bins, range = range)\n",
    "    H = H.T\n",
    "\n",
    "    return H, range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cell_density(df_cells,10000,12500,10000,12500,image=None,bins = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cell_density(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"plot cell density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    H, range = get_cell_density(df_cells,x_min,x_max,y_min,y_max,image,bins)\n",
    "    H = gaussian_filter(H,sigma=smooth)\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(title='cell density') \n",
    "    plot = plt.imshow(H, interpolation='nearest',\n",
    "            extent=[x_min, x_max, y_max, y_min], #norm= mpl.colors.Normalize.__call__(self,value=H)\n",
    "            )\n",
    "    fig.colorbar(plot)\n",
    "\n",
    "    if show_ticks:\n",
    "        ax.tick_params(which='both', direction='in')\n",
    "    else:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cell_density(df_cells,10000,12500,10000,12500,image=None,bins = 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cell_heterogenity(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"get number of different celltypes per bin\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells: pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    result : array of ints\n",
    "        number of different celltypes per bin\n",
    "    range : range of binning    \n",
    "    \"\"\"\n",
    "    if (x_min is None or x_max is None or y_min is None or y_max is None) and image is None:\n",
    "        raise ValueError(\"please provide an image or crop\")         \n",
    "        \n",
    "    if image is None:\n",
    "        range = [[x_min,x_max],[y_min,y_max]]\n",
    "    \n",
    "    else:\n",
    "        range = [[0,image.shape[0]],[0,image.shape[1]]]\n",
    "    \n",
    "    #filter cells\n",
    "    df = df_cells\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "    \n",
    "    \n",
    "    x_bins = np.linspace(x_min,x_max,bins +1)\n",
    "    y_bins = np.linspace(y_min,y_max,bins +1)\n",
    "\n",
    "    df['x_bin'] = np.digitize(df['x'], x_bins, right=True) - 1\n",
    "    df['y_bin'] = np.digitize(df['y'], y_bins, right=True) - 1\n",
    "\n",
    "    grouped = df.groupby(['x_bin', 'y_bin'])['celltype'].apply(lambda x: len(x.unique()))\n",
    "\n",
    "    result = np.zeros((bins, bins))\n",
    "    for (x, y), count in grouped.items():\n",
    "        result[x - 1, y - 1] = count\n",
    "        \n",
    "    return result, range\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cell_heterogenity(df_cells,10000,12500,10000,12500,image=None,bins = 20)  #fix warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cell_heterogenity(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"plot cell density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells: pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    H, range = get_cell_heterogenity(df_cells,x_min,x_max,y_min,y_max,image,bins)\n",
    "    H = gaussian_filter(H,sigma=smooth)\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(title='cell heterogenity') \n",
    "    plot = plt.imshow(H, interpolation='nearest',\n",
    "            extent=[x_min, x_max, y_max, y_min], #norm= mpl.colors.Normalize.__call__(self,value=H)\n",
    "            )\n",
    "    fig.colorbar(plot)\n",
    "\n",
    "    if show_ticks:\n",
    "        ax.tick_params(which='both', direction='in')\n",
    "    else:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cell_heterogenity(df_cells,10000,12500,10000,12500,image=None,bins = 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_celltype_density(df_cells: pd.DataFrame, celltype: str, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"get celltype density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    celltype : str\n",
    "        celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of celltype per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "     \n",
    "    if (x_min is None or x_max is None or y_min is None or y_max is None) and image is None:\n",
    "        raise ValueError(\"please provide an image or crop\")         \n",
    "        \n",
    "    if image is None:\n",
    "        range = [[x_min,x_max],[y_min,y_max]]\n",
    "    \n",
    "    else:\n",
    "        range = [[0,image.shape[0]],[0,image.shape[1]]]\n",
    "    \n",
    "\n",
    "    #filter cells\n",
    "    df = df_cells\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]    #unnÃ¶tig?\n",
    "\n",
    "    H_total, x_edges, y_edges = np.histogram2d(df['x'],df['y'], bins = bins, range = range)\n",
    "\n",
    "    df = df.loc[df[\"celltype\"]==celltype]\n",
    "\n",
    "    H_celltype, x_edges, y_edges = np.histogram2d(df['x'],df['y'], bins = bins, range = range)\n",
    "    H = H_celltype/H_total\n",
    "    H = H.T\n",
    "\n",
    "    return H, range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_celltype_density(df_cells,\"CA3\", 10000,12500,10000,12500,image=None,bins = 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_celltype_density(df_cells: pd.DataFrame, celltype: str, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"plot cell density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells: pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    celltype : str\n",
    "        celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    H, range = get_celltype_density(df_cells,celltype,x_min,x_max,y_min,y_max,image,bins)\n",
    "    H = gaussian_filter(H,sigma=smooth)\n",
    "    x_min, x_max, y_min, y_max = range[0][0], range[0][1], range[1][0], range[1][1]\n",
    "\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(title='celltype density') \n",
    "    plot = plt.imshow(H, interpolation='nearest',\n",
    "            extent=[x_min, x_max, y_max, y_min], #norm= mpl.colors.Normalize.__call__(self,value=H)\n",
    "            )\n",
    "    fig.colorbar(plot)\n",
    "\n",
    "    if show_ticks:\n",
    "        ax.tick_params(which='both', direction='in')\n",
    "    else:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAdAAAAGkCAYAAACB7nTXAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAl30lEQVR4nO3de3RU1fn/8c9MMBMgBNBAAjEaQJSrQEPhF29oVyC2FM2yAkpLMEuhKrFAlopRIVBEvEGpiqIopS5BMVgvrTSU5itaSyyCUm+IWEAQTQhVGgQ1MGf//kBHh1zInNlJZpj3a61Zyzlzzt47JxMfnmfvc47HGGMEAABC4m3pAQAAEI0IoAAAuEAABQDABQIoAAAuEEABAHCBAAoAgAsEUAAAXCCAAgDgAgEUAAAXCKBoEh6PR7NmzQq8X7ZsmTwej3bu3NliY7Jt1qxZ8ng8LT2ME/LcAtGAAIoW89BDD2nZsmUtPYwTEucWaHoEULQY/idvx/jx4/XVV1/p9NNPD2zj3AJNr1VLDwBAeOLi4hQXF9fSwwBiDhkoAvbs2aOrr75aXbt2lc/nU7du3XTdddeppqYmsM/+/fs1depUpaeny+fz6YwzztDdd98tx3FC6isjI0PvvfeeXnnlFXk8Hnk8Hl144YXavn27PB6Pfve739U6Zv369fJ4PHrqqackfT8H+cEHH2jMmDFKSkrSKaecoilTpujrr7+udfyTTz6pzMxMtW7dWieffLKuuOIK7d69u1Hjfe211/TjH/9YCQkJ6tGjhx555JF6921MPxdeeKH69eun999/XxdddJHatGmjtLQ03XPPPbXae+CBB9S3b1+1adNGHTt21ODBg7VixYrA58fOgdo4twCOjwwUkqRPP/1UQ4YM0f79+zVp0iT16tVLe/bs0apVq3To0CHFx8fr0KFDGjZsmPbs2aNf//rXOu2007R+/XoVFRXps88+08KFCxvd38KFC3XDDTcoMTFRt912myQpJSVF3bt317nnnqvly5dr2rRpQccsX75c7dq106WXXhq0fcyYMcrIyNC8efP0+uuv6/7779cXX3yhJ554IrDP3LlzNWPGDI0ZM0bXXHONqqqq9MADD+iCCy7QW2+9pQ4dOtQ71nfeeUcjRoxQp06dNGvWLB05ckTFxcVKSUmptW8o/XzxxRe6+OKLddlll2nMmDFatWqVpk+frv79++unP/2pJGnJkiX6zW9+o8svvzzwD4O3335b//rXvzRu3LgmP7cAGmAAY0xeXp7xer3mjTfeqPWZ4zjGGGPmzJlj2rZtaz788MOgz2+55RYTFxdndu3aFdgmyRQXFwfe/+EPfzCSzI4dOwLb+vbta4YNG1arv0ceecRIMlu2bAlsq6mpMcnJyWbChAmBbcXFxUaSueSSS4KOv/76640k8+9//9sYY8zOnTtNXFycmTt3btB+77zzjmnVqlWt7cfKzc01CQkJ5uOPPw5se//9901cXJz54Z9QKP0MGzbMSDJPPPFEYNs333xjUlNTzS9+8YvAtksvvdT07du3wfE1xbkFcHyUcCHHcfT8889r1KhRGjx4cK3Pv7tUo6SkROeff746duyoffv2BV7Z2dny+/169dVXrYxnzJgxSkhI0PLlywPb1qxZo3379ulXv/pVrf0nT54c9P6GG26QJK1evVqS9Kc//UmO42jMmDFB405NTVXPnj318ssv1zsWv9+vNWvWKDc3V6eddlpge+/evZWTkxO0b6j9JCYmBv088fHxGjJkiLZv3x7Y1qFDB33yySd644036h1jKEI9twDqRwCFqqqqVF1drX79+jW437Zt21RaWqpOnToFvbKzsyVJe/futTKeDh06aNSoUUHzfMuXL1daWpp+8pOf1Nq/Z8+eQe979Oghr9cbmBPctm2bjDHq2bNnrbFv2bKlwXFXVVXpq6++qtWHJJ111llB70Pt59RTT611HWnHjh31xRdfBN5Pnz5diYmJGjJkiHr27KnJkyfrn//8Z73jPZ5Qzy2A+jEHikZzHEfDhw/XzTffXOfnZ555prW+8vLyVFJSovXr16t///568cUXdf3118vrPf6/+Y4NSo7jyOPx6K9//Wudq1UTExOtjDnUfupbOWuMCfx37969tXXrVv3lL39RaWmpnn32WT300EOaOXOmZs+e7Wqc4ZxbAN8jgEKdOnVSUlKS3n333Qb369Gjh7788stAxhmuhu7ic/HFF6tTp05avny5hg4dqkOHDmn8+PF17rtt2zZ169Yt8P6jjz6S4zjKyMgIjNsYo27duoUc5Dt16qTWrVtr27ZttT7bunVr0Ptw+mlI27ZtNXbsWI0dO1Y1NTW67LLLNHfuXBUVFSkhIaHOY2ydWwD145+ckNfrVW5urv785z9r48aNtT7/LiMaM2aMysvLtWbNmlr77N+/X0eOHAmp37Zt22r//v11ftaqVStdeeWVeuaZZ7Rs2TL1799fZ599dp37Llq0KOj9Aw88IEmBlayXXXaZ4uLiNHv27KDs7ruf7b///W+9Y4yLi1NOTo6ef/557dq1K7B9y5Yttc5DOP3U59hj4uPj1adPHxljdPjw4XqPs3VuAdSPDBSSpDvvvFN/+9vfNGzYME2aNEm9e/fWZ599ppKSEr322mvq0KGDbrrpJr344ov6+c9/rquuukqZmZk6ePCg3nnnHa1atUo7d+5UcnJyo/vMzMzUww8/rDvuuENnnHGGOnfuHDQPl5eXp/vvv18vv/yy7r777nrb2bFjhy655BJdfPHFKi8v15NPPqlx48ZpwIABko5mhnfccYeKioq0c+dO5ebmql27dtqxY4eee+45TZo0STfeeGO97c+ePVulpaU6//zzdf311+vIkSOBazPffvvtwH7h9lOXESNGKDU1Veeee65SUlK0ZcsWPfjggxo5cqTatWtX73G2zi2ABrTQ6l9EoI8//tjk5eWZTp06GZ/PZ7p3724mT55svvnmm8A+Bw4cMEVFReaMM84w8fHxJjk52ZxzzjnmvvvuMzU1NYH91IjLWCoqKszIkSNNu3btjKQ6L7vo27ev8Xq95pNPPqn12XeXsbz//vvm8ssvN+3atTMdO3Y0BQUF5quvvqq1/7PPPmvOO+8807ZtW9O2bVvTq1cvM3nyZLN169bjnptXXnnFZGZmmvj4eNO9e3ezePHiQP9u+hk2bFidl6dMmDDBnH766YH3jzzyiLngggvMKaecYnw+n+nRo4e56aabzP/+97/APk1xbgEcn8eYY2pNQAQZNGiQTj75ZJWVldX6bNasWZo9e7aqqqpCynxxVEPnFsDxMQeKiLVx40Zt3rxZeXl5LT2UEw7nFggfc6CIOO+++642bdqk+fPnq0uXLho7dmxLD+mEwbkF7CEDRcRZtWqV8vPzdfjwYT311FP1XqqB0HFuAXuYAwUARLVXX31V9957rzZt2qTPPvtMzz33nHJzcxs8Zt26dSosLNR7772n9PR03X777brqqqtC6pcMFAAQ1Q4ePKgBAwbUuia8Pjt27NDIkSN10UUXafPmzZo6daquueaaOq9xbwgZKADghOHxeI6bgU6fPl0vvfRS0N3XrrjiCu3fv1+lpaWN7qtRi4gcx9Gnn36qdu3aNXiLMABAZDLG6MCBA+ratWuT3ff466+/Vk1NjZW2jDG14o3P55PP5wu77fLy8lq3JM3JydHUqVNDaqdRAfTTTz9Venp6SA0DACLP7t27deqpp1pv9+uvv1a30xNVsddvpb3ExER9+eWXQduKi4s1a9assNuuqKhQSkpK0LaUlBRVV1frq6++UuvWrRvVTqMC6He3DDtPP1MrnRTiUOHGC/97oqWHUMul7SPvmsFIPE9AJKqurlZ6enqDt4AMR01NjSr2+rVj0+lKahdehlt9wFG3zI+1e/duJSUlBbbbyD5talQA/S6NbqWT1MpDAG0OP/zSRIpI/N1H4nkCIllTT8MltfOGHUADbSUlNcnfeGpqqiorK4O2VVZWKikpqdHZp8SNFAAAFvmNI3+YS1P9xrEzmHpkZWVp9erVQdvWrl2rrKyskNrhMhYAgDWOjJVXKL788ktt3rxZmzdvlnT0MpXNmzcHHkFYVFQUdNvKa6+9Vtu3b9fNN9+sDz74QA899JCeeeYZTZs2LaR+CaAAgKi2ceNGDRo0SIMGDZIkFRYWatCgQZo5c6Yk6bPPPgt6nm+3bt300ksvae3atRowYIDmz5+vxx57TDk5OSH1SwkXAGCNI0fhFmBDbeHCCy+s9RD7H1q2bFmdx7z11luhDi0IARQAYI3fGPnDvD9PuMc3F0q4AAC4QAYKALDGzSKgutqIBgRQAIA1joz8BFAAAEITSxkoc6AAALhABgoAsCaWVuESQAEA1jjfvsJtIxpQwgUAwAUyUACANX4Lq3DDPb65EEAj1HDv6JYeAiKAre/BWqfESjvA8fiNLDyNxc5YmholXAAAXCADBQBYE0uLiAigAABrHHnklyfsNqIBJVwAAFwgAwUAWOOYo69w24gGBFAAgDV+CyXccI9vLgRQAIA1sRRAmQMFAMAFMlAAgDWO8cgxYa7CDfP45kIABQBYQwkXAAA0iAwUAGCNX175w8zN/JbG0tQIoAAAa4yFOVATJXOglHABAHCBDBQAYE0sLSIigAIArPEbr/wmzDnQKLmVHyVcAABcIAMFAFjjyCMnzNzMUXSkoFEdQI/8JNNKO63+b5OVdtD8hntHt/QQogLnCUfM4WbphzlQAABcsDMHGh0ZKHOgAAC4QAYKALDm6BxomDeTp4QLAIg1joVb+UXLIiJKuAAAuEAGCgCwJpYWERFAAQDWOPLGzHWglHABAHCBDBQAYI3feOQP83Fk4R7fXAigAABr7DxQmxIuAAAnLDJQAIA1jvHKCXMVrsMqXABArImlEi4BFABgjaPwFwE5dobS5JgDBQDABTJQAIA1dm6kEB25HQEUAGCNnVv5EUCbXNmTj1tpx5u6zUo7kjTcO9paWwCAyBXVARQAEFl4HigAAC7EUgk3OkYJAECEIQMFAFhj50YK0ZHbEUABANY4xiMn3BspRMnTWKIjzAMAEGHIQAEA1jgWSrjcSAEAEHPsPI2FAAoAiDF+eeQP8zrOcI9vLtER5gEAiDBkoAAAayjhAgDggl/hl2D9dobS5KIjzAMAEGHIQAEA1lDCBQDABW4mDwBAFFm0aJEyMjKUkJCgoUOHasOGDQ3uv3DhQp111llq3bq10tPTNW3aNH399dch9UkABQBYY759Hmg4LxPiIqSVK1eqsLBQxcXFevPNNzVgwADl5ORo7969de6/YsUK3XLLLSouLtaWLVv0+OOPa+XKlbr11ltD6pcACgCw5rsSbrivUCxYsEATJ05Ufn6++vTpo8WLF6tNmzZaunRpnfuvX79e5557rsaNG6eMjAyNGDFCV1555XGz1mM1+xzonunnWGurz8N22kqfM9pKO2h+Nr9PaXevt9YWgPBVV1cHvff5fPL5fEHbampqtGnTJhUVFQW2eb1eZWdnq7y8vM52zznnHD355JPasGGDhgwZou3bt2v16tUaP358SONjEREAwBqbjzNLT08P2l5cXKxZs2YFbdu3b5/8fr9SUlKCtqekpOiDDz6os/1x48Zp3759Ou+882SM0ZEjR3TttdeGXMIlgAIArLH5QO3du3crKSkpsP3Y7NOtdevW6c4779RDDz2koUOH6qOPPtKUKVM0Z84czZgxo9HtEEABANbYzECTkpKCAmhdkpOTFRcXp8rKyqDtlZWVSk1NrfOYGTNmaPz48brmmmskSf3799fBgwc1adIk3XbbbfJ6G/cPABYRAQCiVnx8vDIzM1VWVhbY5jiOysrKlJWVVecxhw4dqhUk4+LiJEnGmEb3TQYKALDGkTfsB2KHenxhYaEmTJigwYMHa8iQIVq4cKEOHjyo/Px8SVJeXp7S0tI0b948SdKoUaO0YMECDRo0KFDCnTFjhkaNGhUIpI1BAAUAWOM3HvnDLOGGevzYsWNVVVWlmTNnqqKiQgMHDlRpaWlgYdGuXbuCMs7bb79dHo9Ht99+u/bs2aNOnTpp1KhRmjt3bkj9EkABAFGvoKBABQUFdX62bt26oPetWrVScXGxiouLw+qTAAoAsMbmIqJIRwAFAFhjLDyNxXAzeQAATlxkoAAAa/zyyB/izeDraiMaEEABANY4Jvw5TKfxl2K2KEq4AAC4QAYKALDGsbCIKNzjmwsBFABgzXcPxQ63jWhAAAUAWNMSdyJqKdGRJwMAEGFCykCrV3VXXNvwnseW9rP1YR3/Q9tXDLTWli17pp9jpZ20u+2dp0i01imx0s5w72gr7di05tPN1tr6f5svt9JO+599ZKWdSHUif5+iDXOgAAC44MjCrfyiZA40OsI8AAARhgwUAGCNsbAK10RJBkoABQBYE0tPY6GECwCAC2SgAABrWIULAIALlHABAECDyEABANZwL1wAAFyIpRIuARQAYE0sBVDmQAEAcIEMFABgTSxloARQAIA1sRRAKeECAOACGSgAwBqj8C9DMXaG0uQIoAAAayjhAgCABoWUgSZdvl2tPCc11VhC1n3cZivtHPlJppV2JOn9edOstNMt1bHSjiR12GLnX3PJj5ZbaUeShntHW2lnrVNipR3J3pi8qdustCNJ7X9mZ0wnOlu/O4QvljJQSrgAAGtiKYBSwgUAwAUyUACANbGUgRJAAQDWGOORCTMAhnt8cyGAAgCsiaXHmTEHCgCAC2SgAABrmAMFAMCFWJoDpYQLAIALZKAAAGso4QIA4AIlXAAA0CAyUACANcZCCTdaMlACKADAGiPJhPlE7Gh5oDYlXAAAXCADBQBY48gjT4zcyo8ACgCwJpZW4RJAJbX6v03W2hruHW2lnTOstHLUWqfESjvDH7Xzs0kWx2TpfNsUiWMCmotjPPLEyHWgzIECAOACGSgAwBpjLKzCjZJluARQAIA1sTQHSgkXAAAXyEABANbEUgZKAAUAWMMqXAAA0CAyUACANazCBQDAhaMBNNw5UEuDaWKUcAEAcIEMFABgDatwAQBwwSj853lGSQWXAAoAsCeWMlDmQAEAcIEMFABgTwzVcMlAAQD2fFvCDeclFyXcRYsWKSMjQwkJCRo6dKg2bNjQ4P779+/X5MmT1aVLF/l8Pp155plavXp1SH2SgQIAotrKlStVWFioxYsXa+jQoVq4cKFycnK0detWde7cudb+NTU1Gj58uDp37qxVq1YpLS1NH3/8sTp06BBSvwRQAIA1LXEnogULFmjixInKz8+XJC1evFgvvfSSli5dqltuuaXW/kuXLtXnn3+u9evX66STTpIkZWRkhDxOAmgMGO4d3dJDqCUSxwREorVOiZV2qqur1b59eyttNcTmKtzq6uqg7T6fTz6fL2hbTU2NNm3apKKiosA2r9er7OxslZeX19n+iy++qKysLE2ePFkvvPCCOnXqpHHjxmn69OmKi4tr9DiZAwUARKT09HS1b98+8Jo3b16tffbt2ye/36+UlJSg7SkpKaqoqKiz3e3bt2vVqlXy+/1avXq1ZsyYofnz5+uOO+4IaXxkoAAAe1wuAqrVhqTdu3crKSkpsPnY7NMtx3HUuXNnPfroo4qLi1NmZqb27Nmje++9V8XFxY1uhwAKALDG5hxoUlJSUACtS3JysuLi4lRZWRm0vbKyUqmpqXUe06VLF5100klB5drevXuroqJCNTU1io+Pb9Q4KeECAOwxll6NFB8fr8zMTJWVlQW2OY6jsrIyZWVl1XnMueeeq48++kiO4wS2ffjhh+rSpUujg6dEAAUARLnCwkItWbJEf/zjH7VlyxZdd911OnjwYGBVbl5eXtAio+uuu06ff/65pkyZog8//FAvvfSS7rzzTk2ePDmkfinhAgCsaYl74Y4dO1ZVVVWaOXOmKioqNHDgQJWWlgYWFu3atUte7/f5Ynp6utasWaNp06bp7LPPVlpamqZMmaLp06eH1C8BFABgVwvciq+goEAFBQV1frZu3bpa27KysvT666+H1SclXAAAXCADBQBYE0uPMyOAAgDs4WksAACgIWSgAACLPN++wm0j8hFAAQD2UMIFAAANIQMFANgTQxkoARQAYI/Fp7FEOgIoAMAam09jiXQEUAAhWeuUtPQQmsxw7+iWHkIttsZ0xBy20g6+RwAFANjDHCgAAC7E0Bwol7EAAOACGSgAwBqPOfoKt41oQAAFANgTQ3OglHABAHCBDBQAYE8MLSIigAIA7KGECwAAGkIGCgCwJ4YyUAIoAMAeAigAAC7E0CIi5kABAHCBDBQAYA13IgIAwI0YmgOlhAsAgAsEUAAAXKCECwCwxiMLc6BWRtL0CKDAt9Y6JS09hCYz3Du6pYdQi80x2frd2fwOROI5h10EUACAPTF0HSgBFABgD6twAQBAQ8hAAQD2xFAGSgAFAFjDnYgAAHAjhjJQ5kABAHCBDBQAYE8MZaAEUACANbE0B0oJFwAAF8hAAQD2cCciAABciKE5UEq4AAC4QAYKALAmlhYREUABAPZQwgUAAA0hAwUA2GOhhBstGWizB1Ce+N78bJ1zm+c7EscUid8nW+fJ5t+dLSf6/wsi7ZxXV1erffv2Td9RDJVwyUABAPbEUABlDhQAABfIQAEA1sTSZSxkoAAAuEAABQDABUq4AAB7YmgREQEUAGANc6AAAKBBZKAAALuiJIMMFwEUAGBPDM2BUsIFAMAFMlAAgDWxtIiIAAoAsCeGSrgEUACANbGUgTIHCgCACwRQAIA9xtIrRIsWLVJGRoYSEhI0dOhQbdiwoVHHPf300/J4PMrNzQ25TwIoAMCeFgigK1euVGFhoYqLi/Xmm29qwIABysnJ0d69exs8bufOnbrxxht1/vnnh9bhtwigAICotmDBAk2cOFH5+fnq06ePFi9erDZt2mjp0qX1HuP3+/XLX/5Ss2fPVvfu3V31G9WLiNY6JS09hKgw3DvaSjuReL5tjsnWeULzs/U9iMTvgK0xHTGHrbRzPDYXEVVXVwdt9/l88vl8Qdtqamq0adMmFRUVBbZ5vV5lZ2ervLy83j5++9vfqnPnzrr66qv1j3/8w9U4yUABAPZYLOGmp6erffv2gde8efNqdbdv3z75/X6lpKQEbU9JSVFFRUWdQ3zttdf0+OOPa8mSJWH9qFGdgQIATly7d+9WUlJS4P2x2acbBw4c0Pjx47VkyRIlJyeH1RYBFABgj8UbKSQlJQUF0LokJycrLi5OlZWVQdsrKyuVmppaa////Oc/2rlzp0aNGhXY5jiOJKlVq1baunWrevTo0ahhUsIFAFjz3RxouK/Gio+PV2ZmpsrKygLbHMdRWVmZsrKyau3fq1cvvfPOO9q8eXPgdckll+iiiy7S5s2blZ6e3ui+yUABAFGtsLBQEyZM0ODBgzVkyBAtXLhQBw8eVH5+viQpLy9PaWlpmjdvnhISEtSvX7+g4zt06CBJtbYfDwEUAGBPC9wLd+zYsaqqqtLMmTNVUVGhgQMHqrS0NLCwaNeuXfJ67RdcCaAAAGta6l64BQUFKigoqPOzdevWNXjssmXLQu9QBFAAgE0x9DQWFhEBAOACGSgAwJ4YykAJoAAAazzfvsJtIxpQwgUAwAUyUACAPZRwAQAIXUtdxtISKOECAOACGSgAwB5KuAAAuBQlATBcBFA02nDv6JYeQpNa65S09BCiQiR+D2z97iLxO2BrTNXV1Wrfvr2VtnAUARQAYE0sLSIigAIA7GEOFACA0MVSBsplLAAAuEAGCgCwhxIuAACho4QLAAAaRAYKALCHEi4AAC7EUAClhAsAgAtkoAAAa2JpEREBFABgDyVcAADQEDJQAIA1HmPkMeGlkOEe31wIoAAAe2KohEsABQBYE0uLiJgDBQDABTJQAIA9lHDh1nDv6JYeQpNZ65S09BCaFL+75m3L5vm21VYkfsdt/WxHzGEr7RwPJVwAANAgMlAAgD2UcAEACB0lXAAA0CAyUACAPZRwAQBwJ1pKsOGihAsAgAtkoAAAe4w5+gq3jShAAAUAWBNLq3AJoAAAe2JoERFzoAAAuEAGCgCwxuMcfYXbRjQggAIA7KGECwAAGkIGCgCwhlW4AAC4EUPXgVLCBQDAhajOQG0+0T7SrHVKWnoItdg837Z+vkgck022fr4T/TzZwnkKHyVcAADcYBUuAABoCBkoAMAaSrgAALgRQ6twCaAAAGtiKQNlDhQAABfIQAEA9sTQKlwCKADAGkq4AACgQWSgAAB7HHP0FW4bUYAACgCwJ4bmQCnhAgDgAhkoAMAajywsIrIykqZHAAUA2BNDdyKihAsAgAsEUACANd9dBxruK1SLFi1SRkaGEhISNHToUG3YsKHefZcsWaLzzz9fHTt2VMeOHZWdnd3g/vUhgAIA7DGWXiFYuXKlCgsLVVxcrDfffFMDBgxQTk6O9u7dW+f+69at05VXXqmXX35Z5eXlSk9P14gRI7Rnz56Q+iWAAgCs8Rhj5RWKBQsWaOLEicrPz1efPn20ePFitWnTRkuXLq1z/+XLl+v666/XwIED1atXLz322GNyHEdlZWUh9RvSIqIX/veEkpKSQurgWMO9o8M6/ofWOiVW2rE5Jlsi8TzZZOvni8SfLRLZPE+R+PcSiThP4auurg567/P55PP5grbV1NRo06ZNKioqCmzzer3Kzs5WeXl5o/o5dOiQDh8+rJNPPjmk8ZGBAgDscSy9JKWnp6t9+/aB17x582p1t2/fPvn9fqWkpARtT0lJUUVFRaOGPH36dHXt2lXZ2dkh/ahcxgIAsMZNCbauNiRp9+7dQVXPY7NPG+666y49/fTTWrdunRISEkI6lgAKAIhISUlJx502TE5OVlxcnCorK4O2V1ZWKjU1tcFj77vvPt111136+9//rrPPPjvk8VHCBQDY08yrcOPj45WZmRm0AOi7BUFZWVn1HnfPPfdozpw5Ki0t1eDBg0P4Ab9HBgoAsKcF7kRUWFioCRMmaPDgwRoyZIgWLlyogwcPKj8/X5KUl5entLS0wBzq3XffrZkzZ2rFihXKyMgIzJUmJiYqMTGx0f0SQAEAUW3s2LGqqqrSzJkzVVFRoYEDB6q0tDSwsGjXrl3yer8vuD788MOqqanR5ZdfHtROcXGxZs2a1eh+CaAAAGvc3kno2DZCVVBQoIKCgjo/W7duXdD7nTt3ht5BHQigAAB7uJk8AABoCBkoAMAaj3P0FW4b0YAACgCwJ4ZKuARQAIA9Lp6mUmcbUYA5UAAAXCADBQBYY/NeuJGOAAoAsCeG5kAp4QIA4AIZKADAHqPA8zzDaiMKNHsAXeuUNHeXxxWJY4pEnKfGGe4d3dJDiDm2vps2f3eROKbmEEtzoJRwAQBwgRIuAMAeIwuLiKyMpMkRQAEA9rAKFwAANIQMFABgjyPJY6GNKEAABQBYE0urcAmgAAB7mAMFAAANIQMFANgTQxkoARQAYE8MBVBKuAAAuEAGCgCwh8tYAAAIXSxdxkIJFwAAF8hAAQD2xNAiIgIoAMAex0ieMAOgEx0BlBIuAAAukIECAOyhhNt0hntHN3eXx7XWKbHWViT+fCcym7+7SBSJP18kjikS/+4icUzNw0IAjZInapOBAgDsiaEMlDlQAABcIAMFANjjGIVdgo2SVbgEUACAPcY5+gq3jShACRcAABfIQAEA9sTQIiICKADAnhiaA6WECwCAC2SgAAB7KOECAOCCkYUAamUkTY4SLgAALpCBAgDsoYQLAIALjiMpzBshONFxIwUCKADAnhjKQJkDBQDABTJQAIA9MZSBEkABAPbE0J2Imj2A2nyava0nvp/oT463dc4j8TxF4phsfsdPZDZ/d5F4ziPxuwm7yEABANYY48iE+TiycI9vLgRQAIA9xoRfgo2SOVBW4QIA4AIZKADAHmNhEVGUZKAEUACAPY4jecKcw4ySOVBKuAAAuEAGCgCwhxIuAAChM44jE2YJl8tYAACxJ4YyUOZAAQBwgQwUAGCPYyRPbGSgBFAAgD3GKOwHakdJAKWECwCAC2SgAABrjGNkwizhGjJQAEDMMY6dV4gWLVqkjIwMJSQkaOjQodqwYUOD+5eUlKhXr15KSEhQ//79tXr16pD7JIACAKLaypUrVVhYqOLiYr355psaMGCAcnJytHfv3jr3X79+va688kpdffXVeuutt5Sbm6vc3Fy9++67IfVLAAUAWGMcY+UVigULFmjixInKz89Xnz59tHjxYrVp00ZLly6tc//f//73uvjii3XTTTepd+/emjNnjn70ox/pwQcfDKlfAigAwJ5mLuHW1NRo06ZNys7ODmzzer3Kzs5WeXl5nceUl5cH7S9JOTk59e5fn0YtIvpuQre6ujqkxpvas/sfb+khRAVbvzfOd+NE2t9JpLL5fYrEcx5pfy/V1dVKT3+hyRfoHNHhsG9EdESHJdX+vfp8Pvl8vqBt+/btk9/vV0pKStD2lJQUffDBB3W2X1FRUef+FRUVIY2zUQH0wIEDkqT09PSQGgcARJYDBw6offv21tuNj49XamqqXqsIfTFOXRITE2vFnOLiYs2aNctK+zY0KoB27dpVu3fvVrt27eTxeJp6TAAAy4wxOnDggLp27dok7SckJGjHjh2qqamx0p4xpla8OTb7lKTk5GTFxcWpsrIyaHtlZaVSU1PrbDs1NTWk/evTqADq9Xp16qmnhtQwACCyNEXm+UMJCQlKSEho0j6OFR8fr8zMTJWVlSk3N1eS5DiOysrKVFBQUOcxWVlZKisr09SpUwPb1q5dq6ysrJD65kYKAICoVlhYqAkTJmjw4MEaMmSIFi5cqIMHDyo/P1+SlJeXp7S0NM2bN0+SNGXKFA0bNkzz58/XyJEj9fTTT2vjxo169NFHQ+qXAAoAiGpjx45VVVWVZs6cqYqKCg0cOFClpaWBhUK7du2S1/v9RSfnnHOOVqxYodtvv1233nqrevbsqeeff179+vULqV+PiZZ7JgEAEEG4DhQAABcIoAAAuEAABQDABQIoAAAuEEABAHCBAAoAgAsEUAAAXCCAAgDgAgEUAAAXCKAAALhAAAUAwAUCKAAALvx/dIiyZhADoVMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_celltype_density(df_cells,\"CA3\", 10000,12500,10000,12500,image=None,bins = 20)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "txsim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
