{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import anndata as ad\n",
    "from anndata import AnnData\n",
    "import matplotlib.pyplot as plt\n",
    "from tifffile import imread\n",
    "from scipy.ndimage import gaussian_filter\n",
    "from scipy.sparse import isspmatrix\n",
    "from scipy import stats\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from scipy.sparse import issparse\n",
    "from typing import Tuple\n",
    "from typing import List\n",
    "from skimage.morphology import convex_hull_image \n",
    "import time\n",
    "import ot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_exp0 = ad.read_h5ad('C:/Users/mdichgan/Documents/Helmholtz/send_to_Jakob/spatial/counts_CPc_exp0_BA28.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Yao = ad.read_h5ad(\n",
    "    'C:/Users/mdichgan/Documents/Helmholtz/send_to_Jakob/sc/Yao_150kcells_subsample_with_annotations_sparse_subset.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Yao.obs[\"celltype\"] = adata_Yao.obs[\"label\"]\n",
    "adata_exp0.layers[\"raw\"] = adata_exp0.X\n",
    "adata_Yao.layers[\"raw\"] = adata_Yao.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata_exp0)\n",
    "sc.pp.normalize_total(adata_Yao)\n",
    "adata_exp0.layers[\"lognorm\"] = adata_exp0.X\n",
    "adata_Yao.layers[\"lognorm\"] = adata_Yao.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#helper function\n",
    "def check_crop_exists(x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray):\n",
    "    \"\"\"Check if crop coordinates exist.\n",
    "    \n",
    "    For this, we check if either (x_min, x_max, y_min, y_max) or an image was provided. If not, we raise a ValueError. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    x_min: int, x_max: int, y_min: int, y_max: int\n",
    "        crop coordinates\n",
    "    image: np.ndarray\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    if no ValueError was raised, returns range  \n",
    "    \"\"\"\n",
    "    if (x_min is None or x_max is None or y_min is None or y_max is None) and image is None:\n",
    "        raise ValueError(\"please provide an image or crop\")         \n",
    "        \n",
    "    if x_min is not None and x_max is not None and y_min is not None and y_max is not None:\n",
    "        range = [[x_min,x_max],[y_min,y_max]]\n",
    "    \n",
    "    else:\n",
    "        range = [[0,image.shape[0]],[0,image.shape[1]]]\n",
    "        \n",
    "    return range "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#helper function\n",
    "def matrix_colorbar_plot(matrix: np.ndarray, title: str, x_min: int, x_max: int, y_min: int, y_max: int, vmin, vmax, \n",
    "                         smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Display (smoothed and cropped) matrix as an image with a colorbar and title.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    matrix: np.ndarray\n",
    "        data\n",
    "    title: str\n",
    "    x_min: int, x_max: int, y_min: int, y_max: int\n",
    "        crop coordinates\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    \n",
    "    matrix = gaussian_filter(matrix,sigma=smooth)\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(title = title)\n",
    "    plot = plt.imshow(matrix, vmin=vmin, vmax=vmax, interpolation='nearest', extent=[x_min, x_max, y_max, y_min])\n",
    "    fig.colorbar(plot)\n",
    "    \n",
    "    if not show_ticks:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mean_raster_score_from_cell_score(adata_sp: AnnData, score: str, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "\n",
    "    #optionen lassen bzgl mean, sum etc. ?\n",
    "\n",
    "    df = adata_sp.obs\n",
    "    \n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    #filter cells\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "    bins_x = np.digitize(df['x'], np.linspace(x_min, x_max, bins[0] + 1)) -1\n",
    "    bins_y = np.digitize(df['y'], np.linspace(y_min, y_max, bins[1] + 1)) -1\n",
    "\n",
    "    groups = df.groupby([bins_x, bins_y])\n",
    "    mean = groups[score].mean()\n",
    "    mean_raster_score = np.zeros((bins[0],bins[1]))\n",
    "\n",
    "    for (i,j), value in mean.items():\n",
    "        mean_raster_score[i,j] = value\n",
    "    \n",
    "    return mean_raster_score.T, range\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata_exp0.obs\n",
    "x_min, x_max, y_min, y_max = 11000, 12000, 12000, 11000\n",
    "df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter cells\n",
    "bins_x = np.digitize(df['x'], np.linspace(x_min, x_max,  + 1)) -1\n",
    "bins_y = np.digitize(df['y'], np.linspace(y_min, y_max, 20 + 1)) -1\n",
    "bins_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_eligible_celltypes(adata_sp: AnnData, adata_sc: AnnData, key: str='celltype'):\n",
    "    # Set threshold parameters\n",
    "    min_number_cells=10 # minimum number of cells belonging to a cluster to consider it in the analysis\n",
    "    \n",
    "    #check that genes in spatial data is subset of genes in single cell data\n",
    "    adata_sp = adata_sp[:,adata_sp.var_names.isin(adata_sc.var_names)]\n",
    "\n",
    "    # Subset adata_sc to genes of spatial data\n",
    "    adata_sc = adata_sc[:,adata_sp.var_names]\n",
    "    \n",
    "    # TMP fix for sparse matrices, ideally we don't convert, and instead have calculations for sparse/non-sparse\n",
    "    for a in [adata_sc, adata_sp]:\n",
    "        if issparse(a.layers[\"raw\"]):\n",
    "            a.layers[\"raw\"] = a.layers[\"raw\"].toarray()\n",
    "    \n",
    "    # Get cell types that we find in both modalities\n",
    "    shared_celltypes = adata_sc.obs.loc[adata_sc.obs[key].isin(adata_sp.obs[key]),key].unique()\n",
    "    \n",
    "    # Filter cell types by minimum number of cells\n",
    "    celltype_count_sc = adata_sc.obs[key].value_counts().loc[shared_celltypes]\n",
    "    celltype_count_sp = adata_sp.obs[key].value_counts().loc[shared_celltypes]      #before:adata_sc...\n",
    "    ct_filter = (celltype_count_sc >= min_number_cells) & (celltype_count_sp >= min_number_cells)\n",
    "    celltypes = celltype_count_sc.loc[ct_filter].index.tolist()\n",
    "\n",
    "    return celltypes, adata_sp, adata_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_eligible_celltypes(adata_exp0, adata_Yao)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_negative_marker_dict(adata_sp: AnnData, adata_sc: AnnData, key: str='celltype'):\n",
    "    \"\"\"Add dictionary of negative marker genes for different celltypes.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    key : str\n",
    "        Celltype key in adata_sp.obs and adata_sc.obs\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    neg_marker_dict : dict[string: list]\n",
    "        Dictionary with celltypes as keys and corresponding negative marker genes as values\n",
    "    \"\"\"\n",
    "    max_ratio_cells=0.005     # maximum ratio of cells expressing a marker to call it a negative marker gene-ct pair\n",
    "    adata_sc.layers[\"raw\"] = adata_sc.X\n",
    "    adata_sp.layers[\"raw\"] = adata_sp.X\n",
    "\n",
    "    celltypes, adata_sp, adata_sc  = get_eligible_celltypes(adata_sp, adata_sc, key)  \n",
    "    \n",
    "    # Filter cells to eligible cell types\n",
    "    adata_sc = adata_sc[adata_sc.obs[key].isin(celltypes)]\n",
    "    \n",
    "    # Get ratio of positive cells per cell type\n",
    "    pos_exp_sc = pd.DataFrame(adata_sc.layers[\"raw\"] > 0,columns=adata_sc.var_names)     \n",
    "    pos_exp_sc['celltype'] = list(adata_sc.obs[key])\n",
    "    ratio_celltype_sc = pos_exp_sc.groupby('celltype').mean()\n",
    "    \n",
    "    # Get gene-cell type pairs with negative marker expression\n",
    "    neg_marker_mask = np.array(ratio_celltype_sc < max_ratio_cells)\n",
    "    \n",
    "\n",
    "    neg_marker_mask = pd.DataFrame(neg_marker_mask, columns = adata_sc.var_names, index = ratio_celltype_sc.index)   \n",
    "    neg_marker_dict = {celltype: list(row.index[row]) for celltype, row in neg_marker_mask.iterrows()}\n",
    "\n",
    "    return neg_marker_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_exp0 = get_negative_marker_dict(adata_sp = adata_exp0, adata_sc = adata_Yao, key ='celltype')\n",
    "dict_exp0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spot_assignment_col(adata_sp: AnnData, adata_sc: AnnData):\n",
    "\n",
    "    \"\"\"Add column spot_assignment to adata_sp.uns[\"spots\"].\n",
    "\n",
    "       spot_assignment is \"spot in wrong celltype\", if celltype of cell is in the keys of the neg_marker_dict and gene is a negative marker for the celltype. \n",
    "       spot_assignment is \"unassigned\" if the corresponding cell entry is NaN.\n",
    "       spot_assignment is \"no negative marker\" if gene is  not a negative marker for any celltype in neg_marker_dict.\n",
    "       spot_assignment is \"spot in correct celltype\" otherwise.\n",
    "      \n",
    "       Parameters\n",
    "       ----------\n",
    "       adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "       adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    \"\"\" \n",
    "    #check that genes in spatial data is subset of genes in single cell data\n",
    "    adata_sp = adata_sp[:,adata_sp.var_names.isin(adata_sc.var_names)]\n",
    "\n",
    "    df = adata_sp.uns[\"spots\"]  \n",
    "   \n",
    "    neg_marker_dict = get_negative_marker_dict(adata_sp, adata_sc)\n",
    "\n",
    "    celltypes = list(neg_marker_dict.keys())   \n",
    "\n",
    "    df[\"spot_assignment\"] = \"spot in correct celltype\"\n",
    "\n",
    "    all_neg_markers = set(itertools.chain(*neg_marker_dict.values()))\n",
    "    df.loc[(~df[\"Gene\"].isin(all_neg_markers))& (~df[\"celltype\"].isin(celltypes)), \"spot_assignment\"] = \"no negative marker\"  \n",
    "    #gene is not a negative marker for any celltype in neg_marker_dict and celltype is not in celltypes \n",
    "    for ct in celltypes:\n",
    "      df.loc[(df[\"Gene\"].isin(neg_marker_dict[ct])) & (df[\"celltype\"] == ct), 'spot_assignment'] = \"spot in wrong celltype\" \n",
    "\n",
    "    df.loc[df[\"cell\"]!=df[\"cell\"],\"spot_assignment\"] = \"unassigned\"   #check for NaNs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test get_spot_assignment_col\n",
    "get_spot_assignment_col(adata_sp=adata_exp0, adata_sc=adata_Yao)\n",
    "df_exp0 = adata_exp0.uns[\"spots\"]\n",
    "print(df_exp0['x'].min(), df_exp0['x'].max(), df_exp0['y'].min(), df_exp0['y'].max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"spot in correct celltype\"])/len(df_exp0)\n",
    "b = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"spot in wrong celltype\"])/len(df_exp0)\n",
    "c = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"unassigned\"])/len(df_exp0)\n",
    "d = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"no negative marker\"])/len(df_exp0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(a,b,c,d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cells_location(adata_sp: AnnData, adata_sc: AnnData):\n",
    "    \"\"\"Add x,y coordinate columns of cells to adata_sp.obs\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    \"\"\"\n",
    "    get_spot_assignment_col(adata_sp,adata_sc)\n",
    "    spots = adata_sp.uns[\"spots\"]\n",
    "    df_cells = spots.loc[spots[\"spot_assignment\"]!=\"unassigned\"]      \n",
    "    df_cells = df_cells.groupby([\"cell\"])[[\"x\",\"y\"]].mean()\n",
    "    df_cells = df_cells.reset_index().rename(columns={'cell':'cell_id'})\n",
    "    adata_sp.obs = pd.merge(df_cells,adata_sp.obs,left_on=\"cell_id\",right_on=\"cell_id\",how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cells_location(adata_sp = adata_exp0, adata_sc = adata_Yao)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_knn_mixing_score(adata_st: AnnData, adata_sc: AnnData, obs_key: str = \"celltype\", k: int = 45,ct_filter_factor: float = 2):\n",
    "    \"\"\"Get column in adata_sp.obs with knn mixing score.\n",
    "\n",
    "    For this we concatenate the spatial and single cell datasets, compute the neighborsgraph for eligible celltypes, get the expected value for the\n",
    "    modality ratio, compute the actual ratio for each cell and assign a the knn mixing score.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    \"\"\"\n",
    "    adata_st.obs[\"modality\"] = \"spatial\"\n",
    "    adata_sc.obs[\"modality\"] = \"sc\"\n",
    "    adata = ad.concat([adata_st, adata_sc], join='inner')  \n",
    "    adata.X = adata.X.astype(np.float32)\n",
    "    adata.obs_names = adata.obs_names.astype(str) + \"_\" + adata.obs[\"modality\"].astype(str)\n",
    "    adata.obs = adata.obs.reset_index()\n",
    "\n",
    "    # Set counts to log norm data\n",
    "    adata.X = adata.layers[\"lognorm\"]\n",
    "    \n",
    "    # Calculate PCA (Note: we could also think about pca per cell type...)\n",
    "    assert (adata.obsm is None) or ('X_pca' not in adata.obsm), \"PCA already exists.\"\n",
    "    sc.tl.pca(adata)\n",
    "    \n",
    "    # get cell type groups\n",
    "    sc_cts = set(adata_sc.obs[\"celltype\"].cat.categories)\n",
    "    st_cts = set(adata_st.obs[\"celltype\"].cat.categories)\n",
    "    shared_cts = list(sc_cts.intersection(st_cts))         \n",
    "\n",
    "    # Get ratio per shared cell type\n",
    "    df = pd.DataFrame(columns=[\"index\",\"knn mixing score\"])\n",
    "    for ct in shared_cts:\n",
    "        enough_cells = (adata.obs.loc[adata.obs[obs_key]==ct,\"modality\"].value_counts() > (ct_filter_factor * k)).all()     \n",
    "        if enough_cells:\n",
    "            a = adata[adata.obs[obs_key]==ct]\n",
    "            exp_val = (a.obs.loc[a.obs[\"modality\"]==\"sc\"].shape[0])/a.obs.shape[0]  \n",
    "            sc.pp.neighbors(a,n_neighbors=k)\n",
    "            G = nx.Graph(incoming_graph_data=a.obsp[\"connectivities\"])\n",
    "            nx.set_node_attributes(G, {i:a.obs[\"modality\"].values[i] for i in range(G.number_of_nodes())}, \"modality\")   \n",
    "\n",
    "            ct_df = np.zeros(a.obs.shape[0])\n",
    "            f = np.vectorize(lambda x: x/exp_val if x>=0 and x<=exp_val else x/(exp_val-1)+1/(1-exp_val))\n",
    "            i = 0\n",
    "            for cell in G.nodes():\n",
    "                ct_df[i] = sum(1 for neighbor in G.neighbors(cell) if G.nodes[neighbor][\"modality\"]==\"sc\")  #number_modality_sc\n",
    "                ct_df[i] = ct_df[i]/G.degree(cell)      #ratio: number modality sc / total cells\n",
    "                i += 1 \n",
    "\n",
    "            df_t = a.obs[a.obs['index'].str.endswith('spatial')]    \t    #TODO: früher subsetten\n",
    "            df_t[\"index\"] = df_t[\"index\"].apply(lambda s: s.rsplit('_',1)[0])\n",
    "            ct_df = ct_df[a.obs['index'].str.endswith('spatial')]\n",
    "            \n",
    "            daf = pd.DataFrame({\"index\": df_t[\"index\"], \"knn mixing score\":f(ct_df)})\n",
    "            df = pd.concat([df,daf])\n",
    "\n",
    "    adata_st.obs = pd.merge(adata_st.obs.reset_index(),df,left_on=\"index\",right_on=\"index\",how=\"left\")   \n",
    "         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#vorübergehend random\n",
    "adata_exp0.obs[\"score\"] = np.random.rand((len(adata_exp0.obs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "image = imread(r\"C:\\Users\\mdichgan\\Documents\\Helmholtz\\send_to_Jakob\\tif_images\\background.tiff\")\n",
    "plt.imshow(image)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot cell nuclei allignment\n",
    "df = adata_exp0.obs\n",
    "x_min, x_max, y_max, y_min = 11000, 12000, 12000, 11000\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.axis([x_min, x_max, y_max, y_min])        \n",
    "ax.imshow(image,cmap = \"binary_r\") \n",
    "\n",
    "#filter spots\n",
    "df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "ax.scatter(df['x'], df['y'], s = 0.6, color = \"red\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spots(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, show_ticks: bool = False):\n",
    "    \"\"\"Plot gene spots.\n",
    "     \n",
    "     Spot is red if entry in 'spot_assignment' is \"spot in wrong celltype\", blue if \"spot in correct celltype\", grey if \"unassigned\", green if \"no negative marker\".\n",
    "\n",
    "     Parameters\n",
    "     ----------\n",
    "     adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "     x_min : int, x_max : int, y_min : int, y_max : int \n",
    "          crop coordinates\n",
    "     image : NDArray\n",
    "          read from image of dapi stained cell-nuclei\n",
    "     show_ticks : bool \n",
    "          default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    df = adata_sp.uns[\"spots\"]\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    \n",
    "    s_factor =  150000/((x_max-x_min)**2)                  \n",
    "\n",
    "    plt.axis([x_min, x_max, y_max, y_min])        \n",
    "    plt.imshow(image,cmap = \"binary_r\") \n",
    "\n",
    "    #filter spots\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "    \n",
    "    \n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"unassigned\",\"x\"],df.loc[df['spot_assignment']==\"unassigned\",\"y\"], s = 0.3*s_factor, color = \"grey\", label = \"unassigned\")\n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"spot in correct celltype\",\"x\"],df.loc[df['spot_assignment']==\"spot in correct celltype\",\"y\"], s = 0.5*s_factor, color = \"blue\", label = \"spot in correct celltype\")\n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"spot in wrong celltype\",\"x\"],df.loc[df['spot_assignment']==\"spot in wrong celltype\",\"y\"], s = 1*s_factor, color = \"red\", label = \"spot in wrong celltype\")\n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"no negative marker\",\"x\"],df.loc[df['spot_assignment']==\"no negative marker\",\"y\"], s = 1*s_factor, color = \"green\", label = \"no negative marker\")\n",
    "    \n",
    "    lgnd = plt.legend(loc='center left', bbox_to_anchor=(1, 0.5), scatterpoints=1, markerscale=2, frameon=False)\n",
    "\n",
    "    #fix size of legend spots \n",
    "    lgnd.legend_handles[0]._sizes = [10]\n",
    "    lgnd.legend_handles[1]._sizes = [10]\n",
    "    lgnd.legend_handles[2]._sizes = [10]\n",
    "    lgnd.legend_handles[3]._sizes = [10]\n",
    "\n",
    "    if not show_ticks:\n",
    "        plt.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_spots(adata_exp0, 10000,12000,10000,12000,image,show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_wrong_spot_ratio(adata_sp : AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get ratio array of spots in wrong celltype.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        ratio: spots in wrong celltype over spots in correct celltype or no negative marker spots\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    df = adata_sp.uns[\"spots\"]\n",
    "    np.seterr(invalid='ignore')\n",
    "\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "\n",
    "    true_spots = df.loc[(df['spot_assignment'] == \"spot in correct celltype\") | (df['spot_assignment'] == \"no negative marker\")]\n",
    "    H_t = np.histogram2d(true_spots['x'], true_spots['y'], bins, range)[0].T\n",
    "\n",
    "    spots_in_wrong_ct = df.loc[df['spot_assignment'] == \"spot in wrong celltype\"]\n",
    "    H_w = np.histogram2d(spots_in_wrong_ct['x'], spots_in_wrong_ct['y'], bins, range)[0].T\n",
    "\n",
    "    H = H_w/(H_t+H_w)       #ignore \"Unassigned\" spots in density calculation, since it could be both assigned correctly or wrongly\n",
    "    H[np.isnan(H)] = 0      #no negative marker wrongly assigned where no spots\n",
    "\n",
    " \n",
    "    return H,range\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A,r = get_wrong_spot_ratio(adata_exp0,10000,12500,10000,12500,image=None,bins = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_wrong_spot_ratio(adata_sp : AnnData, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                       image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False): \n",
    "    \"\"\"Plot density of spots in wrong celltype.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    matrix, range = get_wrong_spot_ratio(adata_sp,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    title = \"spots in wrong celltype density\"\n",
    "    vmin, vmax = [0,1]\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_wrong_spot_ratio(adata_exp0,10000,12500,10000,12500,image=None,bins = 50, smooth=0.5)\n",
    "plot_wrong_spot_ratio(adata_exp0,10000,12500,10000,12500,image=None,bins = 50, show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spot_density(adata_sp : AnnData, spot_types: List[str], x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get density of specified spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\")\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    spot_types: list['str']\n",
    "        specify, which spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\") should be included in calculation\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of specified spottypes reads per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    df = adata_sp.uns[\"spots\"]\n",
    "\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    spots = df.loc[df['spot_assignment'].isin(spot_types)]          \n",
    "    H = np.histogram2d(spots['x'], spots['y'], bins, range)[0].T\n",
    "\n",
    "    return H, range\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_spot_density(adata_exp0,[\"spot in wrong celltype\", \"spot in correct celltype\"],10000,12500,10000,12500,image=None,bins = 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spot_density(adata_sp : AnnData, spot_types: list, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot density of specified spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\").\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    spot_types: list[str]                    \n",
    "        specify, which spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\") should be included in calculation\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    matrix, range = get_spot_density(adata_sp,spot_types,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    spot_types_str = '-/'.join([f\"'{spot}'\" for spot in spot_types])\n",
    "    title = f\"density of {spot_types_str} spots\"        \n",
    "    vmin, vmax = [0,None]\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_spot_density(adata_exp0,[\"spot in correct celltype\"],2000,10000,2000,10000,image=None,bins = 50, show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cell_density(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get cell density.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of cells per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "     \n",
    "    df_cells = adata_sp.obs \n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    H = np.histogram2d(df_cells['x'], df_cells['y'], bins, range)[0].T\n",
    "\n",
    "    return H, range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cell_density(adata_exp0,10000,12500,10000,12500,image=None,bins = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cell_density(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot cell density.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    matrix, range = get_cell_density(adata_sp,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    title = \"cell density\"\n",
    "    \n",
    "    vmin, vmax = [0,None]\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cell_density(adata_exp0,11500,12000,11500,12000,image=None,bins = 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_celltype_density(adata_sp: AnnData, celltype: str, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get celltype density.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    celltype: str\n",
    "        celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of celltype per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    df =  adata_sp.obs\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    H_total = np.histogram2d(df['x'],df['y'], bins, range)[0]\n",
    "\n",
    "    df = df.loc[df[\"celltype\"]==celltype]\n",
    "\n",
    "    H_celltype = np.histogram2d(df['x'],df['y'], bins, range)[0]\n",
    "\n",
    "    H = H_celltype/H_total\n",
    "    H[np.isnan(H)] = 0     #0 cells therefore also 0% celltype in the respective area\n",
    "    H = H.T\n",
    "    \n",
    "    return H, range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_celltype_density(adata_exp0,\"CA3\", 10000,12500,10000,12500,image=None,bins = [10,20])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_celltype_density(adata_sp: AnnData, celltypes: list, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot cell density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    celltypes : list[str]\n",
    "        list of celltypes\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    number_of_celltypes = len(celltypes)\n",
    "    \n",
    "    for i in range(number_of_celltypes):\n",
    "        matrix, range_ = get_celltype_density(adata_sp,celltypes[i],x_min,x_max,y_min,y_max,image,bins)\n",
    "        x_min, x_max, y_min, y_max = np.ravel(range_).tolist()\n",
    "        title = f\"{celltypes[i]} density\"\n",
    "        vmin, vmax = [0,None]\n",
    "        matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_celltype_density(adata_exp0,[\"CA3\",\"Cck INH\",\"L6b CTX\",\"Pvalb\"], 2000,12500,2000,12500,image=None,bins = [40,40],show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = np.zeros((2,2,3))\n",
    "a[0,...] = np.array([[1.0,1,1],[1,1,1]])\n",
    "a[1,...] = np.array([[0,1,0],[2,0,3]])\n",
    "print(a)\n",
    "b = np.max(a, axis = 0)\n",
    "print(b)\n",
    "c = np.argmax(a,axis=0)\n",
    "print(c)\n",
    "print(np.dtype(c[1,0]))\n",
    "z = ['a','b']\n",
    "\n",
    "result = np.take(z,c)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_number_of_celltypes(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int]):\n",
    "    \"\"\"Get number of celltypes\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        number of celltypes per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    \n",
    "    range_ = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    celltypes = adata_sp.obs[\"celltype\"].unique()  \n",
    "    A = np.zeros((len(celltypes),bins[0],bins[1]))   \n",
    "\n",
    "    for i in range(len(celltypes)):\n",
    "        A[i,...] = get_celltype_density(adata_sp, celltypes[i],x_min,x_max,y_min,y_max,image,bins)[0]\n",
    "    A = sum(A>0)\n",
    "    \n",
    "    return A, range_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_number_of_celltypes(adata_exp0,10000,12500,10000,12500,image=None,bins = [20,20])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_number_of_celltypes(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int], smooth: float = 0, show_ticks: bool = False):    \n",
    "    \"\"\"Plot number of celltypes\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    matrix, range = get_number_of_celltypes(adata_sp,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    title = \"number of celltypes\"\n",
    "    vmin, vmax = [0,None]\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_number_of_celltypes(adata_exp0,10000,12500,10000,12500,image=None,bins = [40,40])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_major_celltype_perc(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int]):\n",
    "    \"\"\"Get major celltype percentage.\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        major celltype percentage\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    range_ = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    celltypes = adata_sp.obs[\"celltype\"].unique()\n",
    "    A = np.zeros((len(celltypes),bins[0],bins[1]))   \n",
    "    \n",
    "    for i in range(len(celltypes)):\n",
    "        A[i,...] = get_celltype_density(adata_sp, celltypes[i],x_min,x_max,y_min,y_max,image,bins)[0]\n",
    "    B = np.max(A,axis=0)\n",
    "\n",
    "    return B, range_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_major_celltype_perc(adata_exp0,10000,12500,10000,12500,image=None,bins = [20,20])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_major_celltype_perc(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int], smooth: float = 0, show_ticks: bool = False):    \n",
    "    \"\"\"Plot major celltype percentage.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    \n",
    "    matrix, range = get_major_celltype_perc(adata_sp,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    title = \"major celltype percentage\"\n",
    "    vmin, vmax = [0,None]\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_major_celltype_perc(adata_exp0,10000,12500,10000,12500,image=None,bins = [40,40])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_summed_cell_area(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int]):\n",
    "    \"\"\"Get summed cell area.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        summed cell area\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    df = adata_sp.obs\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    #filter cells\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "    bins_x = np.digitize(df['x'], np.linspace(x_min, x_max, bins[0] + 1)) -1\n",
    "    bins_y = np.digitize(df['y'], np.linspace(y_min, y_max, bins[1] + 1)) -1\n",
    "\n",
    "    groups = df.groupby([bins_x, bins_y])\n",
    "    sums = groups['area'].sum()\n",
    "    summed_cell_area = np.zeros((bins[0],bins[1]))\n",
    "\n",
    "    for (i,j), value in sums.items():\n",
    "        summed_cell_area[i,j] = value\n",
    "    \n",
    "    return summed_cell_area.T, range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_summed_cell_area(adata_exp0,10000,12500,10000,12500,image=None,bins = [20,20])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_summed_cell_area(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int], smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot summed cell area.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    matrix, range = get_summed_cell_area(adata_sp,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    title = \"summed cell area\"\n",
    "    vmin, vmax = [0,None]\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_summed_cell_area(adata_exp0,10000,12500,10000,12500,image=None,bins = [40,40], show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_avg_knn_mixing(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, \n",
    "                       image: np.ndarray, bins: Tuple[int,int]):\n",
    "    \"\"\"Get average knn mixing score.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        average knn mixing score\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "\n",
    "    df = adata_sp.obs\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    #filter spots\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "    bins_x = np.digitize(df['x'], np.linspace(x_min, x_max, bins[0] + 1)) -1\n",
    "    bins_y = np.digitize(df['y'], np.linspace(y_min, y_max, bins[1] + 1)) -1\n",
    "\n",
    "    groups = df.groupby([bins_x, bins_y])\n",
    "    nanmean = groups['score'].mean()            #?better with np.nanmean()\n",
    "    average_knn_mixing_score = np.zeros((bins[0],bins[1]))\n",
    "\n",
    "    for (i,j), value in nanmean.items():\n",
    "        average_knn_mixing_score[i,j] = value\n",
    "    \n",
    "    return average_knn_mixing_score.T, range\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_avg_knn_mixing(adata_exp0,10000,12500,10000,12500,image=None,bins = [20,20])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_avg_knn_mixing(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int], smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot summed cell area.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    matrix, range = get_avg_knn_mixing(adata_sp,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    title = \"average knn mixing\"\n",
    "    vmin, vmax = [0,1]\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, vmin, vmax, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_avg_knn_mixing(adata_exp0,10000,12500,10000,12500,image=None,bins = [40,40], show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_correlation_matrices(adata_sp: AnnData, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins: Tuple[int,int], celltype: str):\n",
    "    \"\"\"Get pearson and spearman correlation matrices.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp: AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "        pearson and spearman correlation matrices, df with measurements\n",
    "    \"\"\"\n",
    "\n",
    "    M_wrong_spot_ratio = get_wrong_spot_ratio(adata_sp,x_min, x_max, y_min, y_max,image,bins)[0]\n",
    "    M_spot_density = get_spot_density(adata_sp,[\"spot in wrong celltype\"],x_min, x_max, y_min, y_max,image,bins)[0]\n",
    "    M_cell_density = get_cell_density(adata_sp,x_min, x_max, y_min, y_max,image,bins)[0]\n",
    "    M_celltype_density = get_celltype_density(adata_sp,celltype, x_min, x_max, y_min, y_max,image,bins)[0] \n",
    "    M_number_of_celltypes = get_number_of_celltypes(adata_sp,x_min, x_max, y_min, y_max,image,bins)[0]\n",
    "    M_major_celltype_perc = get_major_celltype_perc(adata_sp,x_min, x_max, y_min, y_max,image,bins)[0]\n",
    "    M_summed_cell_area = get_summed_cell_area(adata_sp,x_min, x_max, y_min, y_max,image,bins)[0]\n",
    "    M_avg_knn_mixing = get_avg_knn_mixing(adata_sp,x_min, x_max, y_min, y_max,image,bins)[0]\n",
    "\n",
    "    measurements_df = pd.DataFrame({'wrong spot ratio':M_wrong_spot_ratio.flatten(), 'spot density': M_spot_density.flatten(), \n",
    "                                'cell density': M_cell_density.flatten(), 'celltype density': M_celltype_density.flatten(), 'number of celltypes': M_number_of_celltypes.flatten(),\n",
    "                                'major celltype perc': M_major_celltype_perc.flatten(), 'summed cell area': M_summed_cell_area.flatten(),\n",
    "                                'avg knn mixing': M_avg_knn_mixing.flatten()})\n",
    "    n = measurements_df.shape[1]\n",
    "    spearman_corr_matrix = np.zeros((n,n))\n",
    "    pearson_corr_matrix = np.zeros((n,n))\n",
    "    for i in range(n):\n",
    "        for j in range(n):\n",
    "            spearman_corr_matrix[i,j] = stats.spearmanr(measurements_df.iloc[:,i],measurements_df.iloc[:,j]).statistic\n",
    "            pearson_corr_matrix[i,j] = stats.pearsonr(measurements_df.iloc[:,i],measurements_df.iloc[:,j]).statistic\n",
    "    \n",
    "    return pearson_corr_matrix, spearman_corr_matrix, measurements_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_correlation_matrices(adata_exp0,0,12500,0,17500,image=None,bins = [20,20], celltype = \"CA3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_correlation_matrices_and_pairplot(adata_sp,x_min,x_max,y_min,y_max,image,bins,celltype):\n",
    "    pearson_corr_matrix, spearman_corr_matrix, measurements_df = get_correlation_matrices(adata_sp,x_min,x_max,y_min,y_max,image,bins,celltype)\n",
    "    fig, (ax1, ax2) = plt.subplots(1,2)\n",
    "    ax1.set_title(\"measurements pearson-corrlations\", fontsize = 10)\n",
    "    ax2.set_title(\"measurements spearman-corrlations\", fontsize = 10)\n",
    "    plot1 = ax1.imshow(pearson_corr_matrix)\n",
    "    plot2 = ax2.imshow(spearman_corr_matrix)\n",
    "\n",
    "    labels = [\"wrong spot ratio\",\"spot density\",\"cell density\", \"celltype density\",\n",
    "            \"number of celltypes\",\"major celltype perc\",\"summed cell area\", \"avg knn mixing\"]\n",
    "    for ax in [ax1,ax2]:\n",
    "        ax.set_xticks([0,1,2,3,4,5,6,7],labels)\n",
    "        ax.set_yticks([0,1,2,3,4,5,6,7],labels)\n",
    "        ax.set_xticklabels(labels, rotation=90)\n",
    "        \n",
    "    fig.colorbar(plot1,fraction=0.046, pad=0.04)    \n",
    "    fig.colorbar(plot2,fraction=0.046, pad=0.04)    \n",
    "\n",
    "    plt.subplots_adjust(wspace=1.5)        \n",
    "\n",
    "    sns.pairplot(measurements_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_correlation_matrices_and_pairplot(adata_exp0,0,12500,0,17500,image=None,bins = [20,20], celltype = \"CA3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from skimage.morphology import convex_hull_image "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "image1 = np.zeros((10, 10))\n",
    "image2 = np.zeros((10, 10))\n",
    "\n",
    "dots_cell1 = np.array([[1, 1], [1, 2], [2, 1], [5, 4], [5, 5]])\n",
    "dots_cell2 = np.array([[8, 8], [8, 9], [5,7], [9, 8], [9, 9]])\n",
    "\n",
    "image1[dots_cell1[:, 0], dots_cell1[:, 1]] = 1\n",
    "image2[dots_cell2[:, 0], dots_cell2[:, 1]] = 1\n",
    "\n",
    "s = convex_hull_image(image1)\n",
    "np.sum(s)\n",
    "\n",
    "# for image in [image1, image2]:\n",
    "#     chull = convex_hull_image(image)\n",
    "    \n",
    "#     fig, axes = plt.subplots(1, 2, figsize=(8, 4))\n",
    "#     x = axes.ravel()\n",
    "    \n",
    "#     x[0].set_title('Original picture')\n",
    "#     x[0].imshow(image, cmap=plt.cm.gray)\n",
    "#     x[0].set_axis_off()\n",
    "    \n",
    "#     x[1].set_title('Transformed picture')\n",
    "#     x[1].imshow(chull, cmap=plt.cm.gray)\n",
    "#     x[1].set_axis_off()\n",
    "    \n",
    "#     plt.tight_layout()\n",
    "#     plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "At = np.ones((10,10))\n",
    "At[dots_cell1[:, 0], dots_cell1[:, 1]] = 0\n",
    "At"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "At[dots_cell1] = 0\n",
    "At"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dots_cell1[:, 0], dots_cell1[:, 1]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "txsim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "0177e28d6b1befb957bf45cd3059bda73a45ca5987be779c0bf4fe7c5743d6a8"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
