{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from anndata import AnnData\n",
    "import anndata\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from tifffile import imread\n",
    "from scipy.ndimage import gaussian_filter\n",
    "from scipy.sparse import csr_matrix, isspmatrix\n",
    "import itertools\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_exp0 = anndata.read_h5ad('C:/Users/mdichgan/Documents/Helmholtz/send_to_Jakob/spatial/counts_CPc_exp0_BA28.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Yao = anndata.read_h5ad(\n",
    "    'C:/Users/mdichgan/Documents/Helmholtz/send_to_Jakob/sc/Yao_150kcells_subsample_with_annotations_sparse_subset.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Yao.obs[\"celltype\"] = adata_Yao.obs[\"label\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#helper function\n",
    "def check_crop_exists(x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray):\n",
    "    \"\"\"Check if crop coordinates exist.\n",
    "    \n",
    "    For this, we check if either (x_min, x_max, y_min, y_max) or an image was provided. If not, we raise a ValueError. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    x_min: int, x_max: int, y_min: int, y_max: int\n",
    "        crop coordinates\n",
    "    image: np.ndarray\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    if no ValueError was raised, returns range  \n",
    "    \"\"\"\n",
    "    if (x_min is None or x_max is None or y_min is None or y_max is None) and image is None:\n",
    "        raise ValueError(\"please provide an image or crop\")         \n",
    "        \n",
    "    if x_min is not None and x_max is not None and y_min is not None and y_max is not None:\n",
    "        range = [[x_min,x_max],[y_min,y_max]]\n",
    "    \n",
    "    else:\n",
    "        range = [[0,image.shape[0]],[0,image.shape[1]]]\n",
    "        \n",
    "    return range "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matrix_colorbar_plot(matrix: np.ndarray, title: str, x_min: int, x_max: int, y_min: int, y_max: int, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Display (smoothed and cropped) matrix as an image with a colorbar and title.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    matrix: np.ndarray\n",
    "        data\n",
    "    title: str\n",
    "    x_min: int, x_max: int, y_min: int, y_max: int\n",
    "        crop coordinates\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    \n",
    "    matrix = gaussian_filter(matrix,sigma=smooth)\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(title = title)\n",
    "    plot = plt.imshow(matrix, interpolation='nearest', extent=[x_min, x_max, y_max, y_min])\n",
    "    fig.colorbar(plot)\n",
    "    \n",
    "    if not show_ticks:\n",
    "        ax.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_negative_marker_dict(adata_sp: AnnData, adata_sc: AnnData, key: str='celltype'):\n",
    "    \"\"\"Add dictionary of negative marker genes for different celltypes.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    key : str\n",
    "        Celltype key in adata_sp.obs and adata_sc.obs\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    neg_marker_dict : dict[string: list]\n",
    "        Dictionary with celltypes as keys and corresponding negative marker genes as values\n",
    "    \"\"\"\n",
    "\n",
    "    # Set threshold parameters\n",
    "    min_number_cells=10 # minimum number of cells belonging to a cluster to consider it in the analysis\n",
    "    max_ratio_cells=0.005 # maximum ratio of cells expressing a marker to call it a negative marker gene-ct pair\n",
    "\n",
    "    #check that genes in spatial data is subset of genes in single cell data\n",
    "    adata_sp = adata_sp[:,adata_sp.var_names.isin(adata_sc.var_names)]\n",
    "\n",
    "    # Subset adata_sc to genes of spatial data\n",
    "    adata_sc = adata_sc[:,adata_sp.var_names]           \n",
    "    \n",
    "    adata_sc.layers[\"raw\"] = adata_sc.X\n",
    "\n",
    "    # TMP fix for sparse matrices, ideally we don't convert, and instead have calculations for sparse/non-sparse\n",
    "    if isspmatrix(adata_sc.layers[\"raw\"]):                                  \n",
    "        adata_sc.layers[\"raw\"] = adata_sc.layers[\"raw\"].toarray()\n",
    "\n",
    "    # Get cell types that we find in both modalities\n",
    "    shared_celltypes = adata_sc.obs.loc[adata_sc.obs[key].isin(adata_sp.obs[key]),key].unique()\n",
    "    \n",
    "    # Filter cell types by minimum number of cells\n",
    "    celltype_count_sc = adata_sc.obs[key].value_counts().loc[shared_celltypes]\n",
    "    celltype_count_sp = adata_sc.obs[key].value_counts().loc[shared_celltypes]   \n",
    "    ct_filter = (celltype_count_sc >= min_number_cells) & (celltype_count_sp >= min_number_cells)\n",
    "    celltypes = celltype_count_sc.loc[ct_filter].index.tolist()             \n",
    "    \n",
    "    \n",
    "    # Filter cells to eligible cell types\n",
    "    adata_sc = adata_sc[adata_sc.obs[key].isin(celltypes)]\n",
    "    genes = adata_sc.var_names\n",
    "    \n",
    "    # Get ratio of positive cells per cell type\n",
    "    pos_exp_sc = pd.DataFrame(adata_sc.layers[\"raw\"] > 0,columns=adata_sp.var_names)     \n",
    "    pos_exp_sc['celltype'] = list(adata_sc.obs[key])\n",
    "    ratio_celltype_sc = pos_exp_sc.groupby('celltype').mean()\n",
    "    \n",
    "    # Get gene-cell type pairs with negative marker expression\n",
    "    neg_marker_mask = np.array(ratio_celltype_sc < max_ratio_cells)\n",
    "    \n",
    "\n",
    "    neg_marker_mask = pd.DataFrame(neg_marker_mask, columns = genes, index = ratio_celltype_sc.index)   \n",
    "    neg_marker_dict = {celltype: list(row.index[row]) for celltype, row in neg_marker_mask.iterrows()}\n",
    "\n",
    "    return neg_marker_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 538,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#test get_negative_marker_dict\n",
    "dict_exp0 = get_negative_marker_dict(adata_sp = adata_exp0, adata_sc = adata_Yao, key ='celltype')\n",
    "len(dict_exp0.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Dkk3</td>\n",
       "      <td>512.391380</td>\n",
       "      <td>5606.902438</td>\n",
       "      <td>Ependymal</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Neurod6</td>\n",
       "      <td>2821.699052</td>\n",
       "      <td>5717.587559</td>\n",
       "      <td>CA3</td>\n",
       "      <td>1855.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ipcef1</td>\n",
       "      <td>189.849074</td>\n",
       "      <td>9254.220586</td>\n",
       "      <td>L6 CT CTX</td>\n",
       "      <td>21448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bhlhe40</td>\n",
       "      <td>163.399427</td>\n",
       "      <td>9492.570656</td>\n",
       "      <td>L6 CT CTX</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Car4</td>\n",
       "      <td>2934.547475</td>\n",
       "      <td>4936.246276</td>\n",
       "      <td>Endo</td>\n",
       "      <td>1464.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9451287</th>\n",
       "      <td>Cdh13</td>\n",
       "      <td>12888.275625</td>\n",
       "      <td>8106.121692</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>12965.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9451288</th>\n",
       "      <td>Fign</td>\n",
       "      <td>12322.526209</td>\n",
       "      <td>8205.875478</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>12562.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9451289</th>\n",
       "      <td>Zfp536</td>\n",
       "      <td>13684.567552</td>\n",
       "      <td>8529.607515</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>20695.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9451290</th>\n",
       "      <td>Cd24a</td>\n",
       "      <td>13119.715081</td>\n",
       "      <td>8672.515345</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>21389.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9451291</th>\n",
       "      <td>Rasgrf2</td>\n",
       "      <td>13720.831688</td>\n",
       "      <td>9328.139451</td>\n",
       "      <td>INH Thal 3</td>\n",
       "      <td>14566.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9451292 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene             x            y    celltype     cell\n",
       "0           Dkk3    512.391380  5606.902438   Ependymal      NaN\n",
       "1        Neurod6   2821.699052  5717.587559         CA3   1855.0\n",
       "2         Ipcef1    189.849074  9254.220586   L6 CT CTX  21448.0\n",
       "3        Bhlhe40    163.399427  9492.570656   L6 CT CTX      NaN\n",
       "4           Car4   2934.547475  4936.246276        Endo   1464.0\n",
       "...          ...           ...          ...         ...      ...\n",
       "9451287    Cdh13  12888.275625  8106.121692       Oligo  12965.0\n",
       "9451288     Fign  12322.526209  8205.875478       Oligo  12562.0\n",
       "9451289   Zfp536  13684.567552  8529.607515       Oligo  20695.0\n",
       "9451290    Cd24a  13119.715081  8672.515345       Oligo  21389.0\n",
       "9451291  Rasgrf2  13720.831688  9328.139451  INH Thal 3  14566.0\n",
       "\n",
       "[9451292 rows x 5 columns]"
      ]
     },
     "execution_count": 539,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_exp0.uns[\"spots\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_neg_mark_read_in_corr_ct(adata_sp: AnnData, adata_sc: AnnData):\n",
    "\n",
    "    \"\"\"Add column in_correct_celltype.\n",
    "\n",
    "       spot_assignment is \"spot in wrong celltype\", if celltype of cell is in the keys of the neg_marker_dict and gene is a negative marker for the celltype. \n",
    "       spot_assignment is \"unassigned\" if the corresponding cell entry is NaN.\n",
    "       spot_assignment is \"no negative marker\" if gene is  not a negative marker for any celltype in neg_marker_dict.\n",
    "       spot_assignment is \"spot in correct celltype\" otherwise.\n",
    "      \n",
    "       Parameters\n",
    "       ----------\n",
    "       adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "       adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "    \n",
    "       Returns\n",
    "       ----------\n",
    "       pandas DataFrame\n",
    "    \"\"\" \n",
    "    #check that genes in spatial data is subset of genes in single cell data\n",
    "    adata_sp = adata_sp[:,adata_sp.var_names.isin(adata_sc.var_names)]\n",
    "\n",
    "    df = adata_sp.uns[\"spots\"]  \n",
    "   \n",
    "    neg_marker_dict = get_negative_marker_dict(adata_sp, adata_sc)\n",
    "\n",
    "    celltypes = list(neg_marker_dict.keys())   \n",
    "\n",
    "    df[\"spot_assignment\"] = \"spot in correct celltype\"\n",
    "\n",
    "    all_neg_markers = set(itertools.chain(*neg_marker_dict.values()))\n",
    "    df.loc[(~df[\"Gene\"].isin(all_neg_markers))& (~df[\"celltype\"].isin(celltypes)), \"spot_assignment\"] = \"no negative marker\"  \n",
    "    #gene is not a negative marker for any celltype in neg_marker_dict and celltype is not in celltypes \n",
    "    for ct in celltypes:\n",
    "      df.loc[(df[\"Gene\"].isin(neg_marker_dict[ct])) & (df[\"celltype\"] == ct), 'spot_assignment'] = \"spot in wrong celltype\" \n",
    "\n",
    "    df.loc[df[\"cell\"]!=df[\"cell\"],\"spot_assignment\"] = \"unassigned\"   #check for NaNs\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 543,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12.245943333879998 13779.88508614 23.3559417021 19435.64276574\n"
     ]
    }
   ],
   "source": [
    "#test get_neg_mark_read_in_corr_ct\n",
    "df_exp0 = get_neg_mark_read_in_corr_ct(adata_sp=adata_exp0, adata_sc=adata_Yao)\n",
    "print(df_exp0['x'].min(), df_exp0['x'].max(), df_exp0['y'].min(), df_exp0['y'].max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "a = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"spot in correct celltype\"])/len(df_exp0)\n",
    "b = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"spot in wrong celltype\"])/len(df_exp0)\n",
    "c = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"unassigned\"])/len(df_exp0)\n",
    "d = len(df_exp0.loc[df_exp0[\"spot_assignment\"]==\"no negative marker\"])/len(df_exp0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.3558064865628953 0.05222989618773814 0.4481002174094293 0.14386339983993723\n"
     ]
    }
   ],
   "source": [
    "print(a,b,c,d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exp0.loc[df_exp0[\"spot_assignment\"]==\"no negative marker\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spots(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, show_ticks: bool = False):\n",
    "    \"\"\"Plot gene spots.\n",
    "     \n",
    "     Spot is red if entry in 'spot_assignment' is \"spot in wrong celltype\", blue if \"spot in correct celltype\", grey if \"unassigned\", green if \"no negative marker\".\n",
    "\n",
    "     Parameters\n",
    "     ----------\n",
    "     spots : pd.DataFrame\n",
    "          pd.DataFrame with columns: gene, x, y, celltype, spot_assignment\n",
    "     x_min : int, x_max : int, y_min : int, y_max : int \n",
    "          crop coordinates\n",
    "     image : NDArray\n",
    "          read from image of dapi stained cell-nuclei\n",
    "     show_ticks : bool \n",
    "          default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    df = spots\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    \n",
    "    s_factor =  150000/((x_max-x_min)**2)                  \n",
    "\n",
    "    plt.axis([x_min, x_max, y_max, y_min])        \n",
    "    plt.imshow(image,cmap = \"binary_r\") \n",
    "\n",
    "    #filter spots\n",
    "    df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "    \n",
    "    \n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"unassigned\",\"x\"],df.loc[df['spot_assignment']==\"unassigned\",\"y\"], s = 0.3*s_factor, color = \"grey\", label = \"unassigned\")\n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"spot in correct celltype\",\"x\"],df.loc[df['spot_assignment']==\"spot in correct celltype\",\"y\"], s = 0.5*s_factor, color = \"blue\", label = \"spot in correct celltype\")\n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"spot in wrong celltype\",\"x\"],df.loc[df['spot_assignment']==\"spot in wrong celltype\",\"y\"], s = 1*s_factor, color = \"red\", label = \"spot in wrong celltype\")\n",
    "    plt.scatter(df.loc[df['spot_assignment']==\"no negative marker\",\"x\"],df.loc[df['spot_assignment']==\"no negative marker\",\"y\"], s = 1*s_factor, color = \"green\", label = \"no negative marker\")\n",
    "    \n",
    "    lgnd = plt.legend(loc='center left', bbox_to_anchor=(1, 0.5), scatterpoints=1, markerscale=2)\n",
    "\n",
    "    #fix size of legend spots \n",
    "    lgnd.legend_handles[0]._sizes = [10]\n",
    "    lgnd.legend_handles[1]._sizes = [10]\n",
    "    lgnd.legend_handles[2]._sizes = [10]\n",
    "    lgnd.legend_handles[3]._sizes = [10]\n",
    "\n",
    "    if not show_ticks:\n",
    "        plt.tick_params(which='both', bottom=False, left=False, labelbottom = False, labelleft = False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "image = imread(r\"C:\\Users\\mdichgan\\Documents\\Helmholtz\\send_to_Jakob\\tif_images\\background.tiff\")\n",
    "plt.imshow(image)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_spots(df_exp0, 11500,12000,11500,12000,image,show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_neg_spot_density(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get density array of spots in wrong celltype.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, spot_assignment\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of negative marker reads per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    df = spots\n",
    "    np.seterr(invalid='ignore')\n",
    "\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "\n",
    "    true_spots = df.loc[(df['spot_assignment'] == \"spot in correct celltype\") | (df['spot_assignment'] == \"no negative marker\")]\n",
    "    H_t = np.histogram2d(true_spots['x'], true_spots['y'], bins, range)[0].T\n",
    "\n",
    "    spots_in_wrong_ct = df.loc[df['spot_assignment'] == \"spot in wrong celltype\"]\n",
    "    H_w = np.histogram2d(spots_in_wrong_ct['x'], spots_in_wrong_ct['y'], bins, range)[0].T\n",
    "\n",
    "    H = H_w/(H_t+H_w)       #ignore \"Unassigned\" spots in density calculation, since it could be both assigned correctly or wrongly\n",
    "    H[np.isnan(H)] = 0      #no negative marker wrongly assigned where no spots\n",
    "\n",
    " \n",
    "    return H,range\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A,r = get_neg_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 20)\n",
    "A.dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_neg_spot_density(spots: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                       image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False): \n",
    "    \"\"\"Plot density of spots in wrong celltype.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, spot_assignment\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "\n",
    "    matrix, range = get_neg_spot_density(spots,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    title = \"spots in wrong celltype density\"\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, smooth, show_ticks)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_neg_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 50, smooth=1)\n",
    "plot_neg_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 50, show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spot_density(spots: pd.DataFrame, spot_types: list['str'], x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get density of specified spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\")\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, spot_assignment\n",
    "    spot_types: list['str']\n",
    "        specify, which spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\") should be included in calculation\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of specified spottypes reads per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    df = spots\n",
    "\n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    true_spots = df.loc[df['spot_assignment'].isin(spot_types)]           #sinnvoll?\n",
    "    H = np.histogram2d(true_spots['x'], true_spots['y'], bins, range)[0].T\n",
    "\n",
    "    return H, range\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_spot_density(df_exp0,[\"spot in wrong celltype\", \"spot in correct celltype\"],10000,12500,10000,12500,image=None,bins = 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spot_density(spots: pd.DataFrame, spot_types: list, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot density of specified spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\").\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    spots : pd.DataFrame\n",
    "        pd.DataFrame with columns: gene, x, y, celltype, spot_assignment\n",
    "    spot_types: list[str]                    \n",
    "        specify, which spot types (\"spot in wrong celltype\", \"spot in correct celltype\", \"unassigned\", \"no negative marker\") should be included in calculation\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    matrix, range = get_spot_density(spots,spot_types,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    title = f\"density of {spot_types} spots\"        #schöner machen\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_spot_density(df_exp0,[\"spot in wrong celltype\", \"spot in correct celltype\"],2000,10000,2000,10000,image=None,bins = 50, show_ticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_exp0.uns[\"spots\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cells_df(adata_sp: AnnData, adata_sc: AnnData):\n",
    "    \"\"\"Get cell location DataFrame.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        Annotated ``AnnData`` object with counts from spatial data\n",
    "    adata_sc : AnnData\n",
    "        Annotated ``AnnData`` object with counts scRNAseq data\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        DataFrame with columns: cell, celltype, x, y\n",
    "    \"\"\"\n",
    "    spots = get_neg_mark_read_in_corr_ct(adata_sp,adata_sc)\n",
    "    df_cells = spots.loc[spots[\"spot_assignment\"]!=\"Unknown\"]      \n",
    "    df_cells = df_cells.groupby([\"cell\",\"celltype\"])[[\"x\",\"y\"]].mean()\n",
    "    df_cells = df_cells.reset_index().rename(columns={'cell':'cell', 'celltype':'celltype'})\n",
    "\n",
    "\n",
    "    return df_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cells = get_cells_df(adata_sp = adata_exp0, adata_sc = adata_Yao)\n",
    "df_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata_exp0.uns[\"spots\"])\n",
    "df_exp0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_exp0.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_cells[\"cell\"].unique())\n",
    "# df_cells.loc[df_cells[\"cell\"]==5.0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_t = pd.merge(df_cells,adata_exp0.obs,left_on=\"cell\",right_on=\"cell_id\",how=\"inner\")      \n",
    "df_t.drop([\"celltype_y\"],axis=1)\n",
    "df_t.rename(columns={\"celltype_x\": \"celltype\"})\n",
    "\n",
    "#TODO: assign it to adata_exp0.obs, ?different lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_cells  \n",
    "x_min, x_max, y_max, y_min = 11000, 12000, 12000, 11000\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.axis([x_min, x_max, y_max, y_min])        \n",
    "ax.imshow(image,cmap = \"binary_r\") \n",
    "\n",
    "\n",
    "#filter spots\n",
    "df = df.loc[(df['x']>= x_min) & (df['x']<=x_max) & (df['y']>=y_min) & (df['y']<=y_max)]\n",
    "\n",
    "ax.scatter(df['x'], df['y'], s = 0.6, color = \"red\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cell_density(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get cell density.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of cells per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "     \n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    H = np.histogram2d(df_cells['x'], df_cells['y'], bins, range)[0].T\n",
    "\n",
    "    return H, range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cell_density(df_cells,10000,12500,10000,12500,image=None,bins = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cell_density(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot cell density.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    matrix, range = get_cell_density(df_cells,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    title = \"cell density\"\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, smooth, show_ticks)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cell_density(df_cells,10000,12500,10000,12500,image=None,bins = 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_celltype_density(df_cells: pd.DataFrame, celltype: str, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get celltype density.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    celltype : str\n",
    "        celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        density of celltype per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    \n",
    "    df = df_cells \n",
    "    range = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    \n",
    "    H_total = np.histogram2d(df['x'],df['y'], bins, range)[0]\n",
    "\n",
    "    df = df.loc[df[\"celltype\"]==celltype]\n",
    "\n",
    "    H_celltype = np.histogram2d(df['x'],df['y'], bins, range)[0]\n",
    "\n",
    "    H = H_celltype/H_total\n",
    "    H[np.isnan(H)] = 0     #0 cells therefore also 0% celltype in the respective area\n",
    "    H = H.T\n",
    "\n",
    "    return H, range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = get_celltype_density(df_cells,\"CA3\", 10000,12500,10000,12500,image=None,bins = [10,20])[0]\n",
    "A.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_celltype_density(df_cells: pd.DataFrame, celltype: str, x_min: int, x_max: int, y_min: int, y_max: int,\n",
    "                      image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):\n",
    "    \"\"\"Plot cell density\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells: pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    celltype : str\n",
    "        celltype\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : int or array_like or [int, int] or [array, array]\n",
    "        The bin specification:\n",
    "        If int, the number of bins for the two dimensions (nx=ny=bins).\n",
    "        If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).\n",
    "        If [int, int], the number of bins in each dimension (nx, ny = bins).\n",
    "        If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).\n",
    "        A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    matrix, range = get_celltype_density(df_cells,celltype,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    title = \"celltype density\"\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, smooth, show_ticks)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_celltype_density(df_cells,\"CA3\", 10000,12500,10000,12500,image=None,bins = 20)\n",
    "#komplettes bild\n",
    "#subplots, liste an celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = np.zeros((2,2,3))\n",
    "a[0,...] = np.array([[1.0,1,1],[1,1,1]])\n",
    "a[1,...] = np.array([[0,1,0],[2,0,3]])\n",
    "print(a)\n",
    "b = np.max(a, axis = 0)\n",
    "print(b)\n",
    "c = np.argmax(a,axis=0)\n",
    "print(c)\n",
    "print(np.dtype(c[1,0]))\n",
    "z = ['a','b']\n",
    "\n",
    "result = np.take(z,c)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_number_of_celltypes(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get number of celltypes\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        number of celltypes per bin\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    \n",
    "    #? celltypes auswählen\n",
    "    #wie bins: [int,int]\n",
    "    \n",
    "    celltypes = df_cells[\"celltype\"].unique()  \n",
    "    A = np.zeros((len(celltypes),bins[0],bins[1]))   #order of axis?\n",
    "\n",
    "    for i in range(len(celltypes)):\n",
    "        A[i,...] = get_celltype_density(df_cells, celltypes[i],x_min,x_max,y_min,y_max,image,bins)[0]\n",
    "    A = sum(A>0)\n",
    "    range_ = check_crop_exists(x_min,x_max,y_min,y_max,image)\n",
    "    return A, range_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_number_of_celltypes(df_cells,10000,12500,10000,12500,image=None,bins = [20,20])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_get_number_of_celltypes(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):    \n",
    "    \"\"\"Plot number of celltypes\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells: pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    matrix, range = get_number_of_celltypes(df_cells,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "    \n",
    "    title = \"number of celltypes\"\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, smooth, show_ticks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_get_number_of_celltypes(df_cells,10000,12500,10000,12500,image=None,bins = [40,40])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_major_celltype_perc(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins):\n",
    "    \"\"\"Get major celltype percentage.\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    df_cells : pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    Returns\n",
    "    -------\n",
    "    H : array of floats\n",
    "        major celltype percentage\n",
    "    range : range of binning \n",
    "    \"\"\"\n",
    "    celltypes = df_cells[\"celltype\"].unique()\n",
    "    A = np.zeros((len(celltypes),bins[0],bins[1]))   #order of axis?\n",
    "    # B = np.zeros((2,bins[0],bins[1]))\n",
    "    \n",
    "    for i in range(len(celltypes)):\n",
    "        A[i,...] = get_celltype_density(df_cells, celltypes[i],x_min,x_max,y_min,y_max,image,bins)[0]\n",
    "    B = np.max(A,axis=0)\n",
    "    range_ = get_celltype_density(df_cells, celltypes[0],x_min,x_max,y_min,y_max,image,bins)[1]\n",
    "    # B[1,...] = np.argmax(A,axis = 0)\n",
    "    # B[1,...] = np.take(celltypes,B[1,...])\n",
    "\n",
    "    return B, range_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_major_celltype_perc(df_cells,10000,12500,10000,12500,image=None,bins = [20,20])[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_get_major_celltype_perc(df_cells: pd.DataFrame, x_min: int, x_max: int, y_min: int, y_max: int, image: np.ndarray, bins, smooth: float = 0, show_ticks: bool = False):    \n",
    "    \"\"\"Plot major celltype percentage.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df_cells: pd.DataFrame\n",
    "        pd.DataFrame with columns: cell, celltype, x, y\n",
    "    x_min : int, x_max : int, y_min : int, y_max : int \n",
    "        crop coordinates\n",
    "    image : NDArray\n",
    "        read from image of dapi stained cell-nuclei\n",
    "    bins : [int,int]\n",
    "        the number of bins in each dimension\n",
    "    smooth : float = 0\n",
    "        sigma parameter of scipy.ndimage.gaussian_filter function\n",
    "    show_ticks : bool \n",
    "        default False, show no ticks or labels\n",
    "    \"\"\"\n",
    "    \n",
    "    matrix, range = get_major_celltype_perc(df_cells,x_min,x_max,y_min,y_max,image,bins)\n",
    "    x_min, x_max, y_min, y_max = np.ravel(range).tolist()\n",
    "\n",
    "    title = \"major celltype percentage\"\n",
    "    matrix_colorbar_plot(matrix, title, x_min, x_max, y_min, y_max, smooth, show_ticks)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_get_major_celltype_perc(df_cells,10000,12500,10000,12500,image=None,bins = [40,40])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_correlate_mats(mat_1: np.ndarray, mat_2: np.ndarray):\n",
    "    \"\"\"Compute the Pearson correlation coefficients between flattened mat_1 and mat_2.\n",
    "       \n",
    "    Parameters\n",
    "    ----------\n",
    "    mat_1: np.ndarray\n",
    "        first matrix\n",
    "    mat_2: np.ndarray\n",
    "        second matrix\n",
    "    Returns\n",
    "    -------\n",
    "    corr_coeff : float   \n",
    "    \"\"\"\n",
    "    mat_1 = mat_1.flatten()\n",
    "    mat_2 = mat_2.flatten()\n",
    "\n",
    "    corr_coeff = np.corrcoef(mat_1,mat_2)[0,1]\n",
    "    \n",
    "    return corr_coeff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = get_neg_spot_density(df_exp0,10000,12500,10000,12500,image=None,bins = 20)[0]\n",
    "B = get_cell_density(df_cells,10000,12500,10000,12500,image=None,bins = 20)[0]\n",
    "get_correlate_mats(A,B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = get_cell_density(df_cells,10000,12500,10000,12500,image=None,bins = 20)[0]\n",
    "B = get_number_of_celltypes(df_cells,10000,12500,10000,12500,image=None,bins = [20,20])[0]\n",
    "get_correlate_mats(A,B)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "txsim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "0177e28d6b1befb957bf45cd3059bda73a45ca5987be779c0bf4fe7c5743d6a8"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
